
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Next Steps &#8212; NVIDIA BioNeMo Framework</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="https://js.hcaptcha.com/1/api.js"></script>
    <script src="//assets.adobedtm.com/5d4962a43b79/c1061d2c5e7b/launch-191c2462b890.min.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script src="_static/design-tabs.js"></script>
    <link rel="canonical" href="https://docs.nvidia.com/bionemo-framework/0.4.0/next-steps.html" />
    <link rel="shortcut icon" href="_static/nvidia-logo-vert-rgb-blk-for-screen.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
      
<style type="text/css">
  ul.ablog-archive {
    list-style: none;
    overflow: auto;
    margin-left: 0px;
  }
  ul.ablog-archive li {
    float: left;
    margin-right: 5px;
    font-size: 80%;
  }
  ul.postlist a {
    font-style: italic;
  }
  ul.postlist-style-disc {
    list-style-type: disc;
  }
  ul.postlist-style-none {
    list-style-type: none;
  }
  ul.postlist-style-circle {
    list-style-type: circle;
  }
</style>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/nvidia-logo-horiz-rgb-blk-for-screen.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">NVIDIA BioNeMo Framework</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GETTING STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="index.html">
   What is BioNeMo?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pre-reqs.html">
   Hardware and Software Prerequisites
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="access-startup.html">
   Access and Startup
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="initialization-guide.html">
   Initialization Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial-list.html">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BioNeMo CORE
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="bionemo-fw-for-model-training-fw.html">
   Fundamentals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="data-module-fw.html">
   Data Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dwnstr-task-validation.html">
   Validation with a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="inference-triton-fw.html">
   Inference with Nvidia Triton Server
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deep-dive-esm1-pytriton-inference.html">
   Example of PyTriton Inference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BEST PRACTICES
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="hyperparameters-fw.html">
   Hyperparameter Usage and Tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="parallelism-fw.html">
   Training Parallelism
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  MODELS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="models/diffdock.html">
   DiffDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/dnabert.html">
   DNABERT
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/dsmbind.html">
   Model Overview
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/equidock.html">
   EquiDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/esm1-nv.html">
   ESM-1nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/esm2-nv.html">
   ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/geneformer.html">
   Geneformer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/megamolbart.html">
   MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/molmim.html">
   MolMIM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/openfold.html">
   OpenFold
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/prott5nv.html">
   Prott5nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/model-benchmarks.html">
   Model Benchmarks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  DATASETS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/uniprot.html">
   UniProt Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/CELLxGENE.html">
   CELLxGENE
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/zinc15.html">
   ZINC-15 Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/pdb.html">
   The Protein Data Bank (PDB)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/GRCh38.p13.html">
   Human Reference Genome Version GRCh38.p13
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/openproteinset.html">
   OpenProteinSet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/moleculenet-physchem.html">
   MoleculeNet Physical Chemistry Datasets - Lipophilicity, FreeSolv, ESOL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  TUTORIALS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="preprocessing-bcp-training-diffdock.html">
   DiffDock: Preparing Workspace and Data for Pre-training
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2nv-mutant-design.html">
   Zero-Shot Protein Design Using ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/MMB_GenerativeAI_Inference_with_examples.html">
   BioNeMo - MegaMolBART Inferencing for Generative Chemistry
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/MolMIM_GenerativeAI_local_inference_with_examples.html">
   MolMIM Inferencing for Generative Chemistry and Downstream Prediction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/ZINC15-data-preprocessing.html">
   ZINC Training Dataset Setup for small molecule language models
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/bionemo-finetuning-overview.html">
   Finetune pre-trained models in BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/cma_es_guided_molecular_optimization_molmim.html">
   MolMIM Property Guided Molecular Optimization Using CMA-ES
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/custom-dataset-class-fw.html">
   Adding the OAS Dataset: Modifying the Dataset Class
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/custom-dataset-dataloader.html">
   Adding the OAS Dataset: Customizing Dataset Object and Dataloader Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/custom-dataset-preprocessing-fw.html">
   Adding the OAS Dataset: Downloading and Preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/encoder-finetuning-notebook-fw.html">
   Encoder Fine-tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_diffdock.html">
   DiffDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_equidock.html">
   EquiDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_esm1nv.html">
   ESM1nv Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_esm2nv.html">
   ESM-2nv: Data Preprocessing and Model Training Using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_mmb.html">
   MegaMolBART Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_molmim.html">
   Train MolMIM from scratch on your own data using the BioNeMo Framework
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/physchem-notebook-fw.html">
   Finetuning LLM in BioNeMo for a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/protein-esm2nv-clustering.html">
   Generating Embeddings for Protein Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/retrosynthesis-notebook.html">
   Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2_FLIP_finetuning.html">
   Fine-Tune ESM-2nv on FLIP Data for Sequence-Level Classification, Regression, Token-Level Classification, and with LoRA Adapters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2_oas_inferencing.html">
   Performing inference on OAS sequences with ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2_paratope_finetuning.html">
   Pretrain from Scratch, Continue Training from an Existing Checkpoint, and Fine-tune ESM-2nv on Custom Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/dnabert_inference.html">
   Generating and visualizing embeddings with DNABert
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/dnabert_pretrain_finetune.html">
   Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  APPENDIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="faq-fw.html">
   Frequently Asked Questions (FAQ)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bibliography-fw.html">
   Bibliography
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="releasenotes-fw.html">
   Release Notes
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ngc-cli-configuration">
   NGC CLI Configuration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#first-time-setup">
   First-Time Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-model-weights">
   Download Model Weights
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#directory-structure">
   Directory Structure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#weights-and-biases-setup-optional">
   Weights and Biases Setup (Optional)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-tutorials">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Next Steps</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#ngc-cli-configuration">
   NGC CLI Configuration
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#first-time-setup">
   First-Time Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-model-weights">
   Download Model Weights
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#directory-structure">
   Directory Structure
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#weights-and-biases-setup-optional">
   Weights and Biases Setup (Optional)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-tutorials">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                 <div class="tex2jax_ignore mathjax_ignore section" id="next-steps">
<h1>Next Steps<a class="headerlink" href="#next-steps" title="Permalink to this headline">#</a></h1>
<p>Now that you have successfully launched the Docker container and entered it, this section will guide you through the container, initial steps to take within the container (such as configuration, downloading pre-trained model weights, etc.), and where to find tutorials.</p>
<div class="section" id="ngc-cli-configuration">
<h2>NGC CLI Configuration<a class="headerlink" href="#ngc-cli-configuration" title="Permalink to this headline">#</a></h2>
<p>NVIDIA NGC Command Line Interface (CLI) is a command-line tool for managing Docker containers in NGC. If NGC is not already installed in the container, download it as per the instructions <a class="reference external" href="https://org.ngc.nvidia.com/setup/installers/cli">here</a> (note that within the container, the AMD64 Linux version should be installed).</p>
<p>Once installed, run <code class="docutils literal notranslate"><span class="pre">ngc</span> <span class="pre">config</span> <span class="pre">set</span></code> to establish NGC credentials within the container.</p>
</div>
<div class="section" id="first-time-setup">
<h2>First-Time Setup<a class="headerlink" href="#first-time-setup" title="Permalink to this headline">#</a></h2>
<p>First, invoke the following launch script. The first time, it will create a .env file and exit:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./launch.sh
</pre></div>
</div>
<p>Next, edit the .env file with the correct NGC parameters for your organization and team:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w">    </span><span class="nv">NGC_CLI_API_KEY</span><span class="o">=</span>&lt;YOUR_API_KEY&gt;
<span class="w">    </span><span class="nv">NGC_CLI_ORG</span><span class="o">=</span>&lt;YOUR_ORG&gt;
<span class="w">    </span><span class="nv">NGC_CLI_TEAM</span><span class="o">=</span>&lt;YOUR_TEAM&gt;
</pre></div>
</div>
</div>
<div class="section" id="download-model-weights">
<h2>Download Model Weights<a class="headerlink" href="#download-model-weights" title="Permalink to this headline">#</a></h2>
<p>You may now download all pre-trained model checkpoints from NGC through the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>./launch.sh<span class="w"> </span>download
</pre></div>
</div>
<p>This will download all models to the <code class="docutils literal notranslate"><span class="pre">workspace/bionemo/models</span></code> directory. Optionally, you may persist the models by copying them to your mounted workspace, so that they need not be redownloaded each time.</p>
</div>
<div class="section" id="directory-structure">
<h2>Directory Structure<a class="headerlink" href="#directory-structure" title="Permalink to this headline">#</a></h2>
<p>Note that <code class="docutils literal notranslate"><span class="pre">workspace/bionemo</span></code> is the home directory for the container. Below are a few key components:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bionemo</span></code>: Contains the core BioNeMo package, which includes base classes for BioNeMo data modules, tokenizers, models, etc.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">examples</span></code>: Contains example scripts, datasets, YAML files, and notebooks</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">models</span></code>: Contains all pre-trained models checkpoints in .nemo format.</p></li>
</ul>
</div>
<div class="section" id="weights-and-biases-setup-optional">
<h2>Weights and Biases Setup (Optional)<a class="headerlink" href="#weights-and-biases-setup-optional" title="Permalink to this headline">#</a></h2>
<p>Training progress and charts of the models can be visualized through <a class="reference external" href="https://docs.wandb.ai/guides/track/public-api-guide">Weights and Biases</a>. Setup your <a class="reference external" href="https://docs.wandb.ai/guides/track/public-api-guide#authentication">API Key</a> to enable logging.</p>
</div>
<div class="section" id="bionemo-framework-tutorials">
<h2>BioNeMo Framework Tutorials<a class="headerlink" href="#bionemo-framework-tutorials" title="Permalink to this headline">#</a></h2>
<p>The best way to get started with BioNeMo Framework is with the tutorials. Below are some of the example walkthroughs which contain code snippets that you can run from within the container.</p>
<p>Tutorials are presented as notebooks (<code class="docutils literal notranslate"><span class="pre">.ipynb</span></code> format), which may contain various code snippets in formats like Python, Bash, YAML, etc. You can follow the instructions in these files, make appropriate code changes, and execute them in the container.</p>
<p>It is convenient to first launch the BioNeMo Framework container and copy the tutorial files to the container, either via the JupyterLab interface drag-and-drop or by mounting the files during the launch of the container (<code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">run</span> <span class="pre">-v</span> <span class="pre">...</span></code>).</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Topic</p></th>
<th class="head"><p>Title</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Model Pre-Training</p></td>
<td><p><a class="reference internal" href="notebooks/model_training_mmb.html"><span class="doc std std-doc">Launching a MegaMolBART model pre-training with ZINC-15 dataset</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Custom Datasets</p></td>
<td><p><a class="reference internal" href="notebooks/ZINC15-data-preprocessing.html"><span class="doc std std-doc">Setting up the ZINC15 dataset used for training MolMIM</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Model Pre-Training</p></td>
<td><p><a class="reference internal" href="notebooks/model_training_molmim.html"><span class="doc std std-doc">Launching a MolMIM model pre-training with ZINC-15 dataset, both from scratch and starting from an existing checkpoint</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Model Pre-Training</p></td>
<td><p><a class="reference internal" href="notebooks/model_training_esm1nv.html"><span class="doc std std-doc">Launching an ESM-1nv model pre-training with UniRef50 dataset</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Model Pre-Training</p></td>
<td><p><a class="reference internal" href="notebooks/model_training_esm2nv.html"><span class="doc std std-doc">Launching an ESM-2nv model pre-training with curated data from UniRef50, UniRef90</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Model Pre-Training</p></td>
<td><p><a class="reference internal" href="notebooks/geneformer_cellxgene_tutorial.html"><span class="doc std std-doc">Pretraining a geneformer model for representing single cell RNA-seq data</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Geneformer Benchmarking</p></td>
<td><p><a class="reference internal" href="notebooks/Geneformer-celltype-classification-example.html"><span class="doc std std-doc">Benchmarking pre-trained Geneformer models against a baseline with cell type classification</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Model Training</p></td>
<td><p><a class="reference internal" href="notebooks/model_training_equidock.html"><span class="doc std std-doc">Launching an EquiDock model pre-training with DIPS or DB5 datasets</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Inference</p></td>
<td><p><a class="reference internal" href="notebooks/MMB_GenerativeAI_Inference_with_examples.html"><span class="doc std std-doc">Performing Inference with MegaMolBART for Generative Chemistry and Predictive Modeling with RAPIDS</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Inference</p></td>
<td><p><a class="reference internal" href="notebooks/MolMIM_GenerativeAI_local_inference_with_examples.html"><span class="doc std std-doc">Performing Inference with MolMIM for Generative Chemistry and Predictive Modeling with RAPIDS</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Inference</p></td>
<td><p><a class="reference internal" href="notebooks/protein-esm1nv-clustering.html"><span class="doc std std-doc">Performing Inference with ESM1-nv and Predictive Modeling with RAPIDS</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Inference</p></td>
<td><p><a class="reference internal" href="notebooks/protein-esm2nv-clustering.html"><span class="doc std std-doc">Performing Inference with ESM2-nv and Predictive Modeling with RAPIDS</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Inference</p></td>
<td><p><a class="reference internal" href="notebooks/cma_es_guided_molecular_optimization_molmim.html"><span class="doc std std-doc">Performing Property-guided Molecular Optimization with MolMIM, which internally involves inference</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Inference</p></td>
<td><p><a class="reference internal" href="notebooks/geneformer_cellxgene_pretrained_inference_tutorial.html"><span class="doc std std-doc">Performing inference and cell clustering on CELLxGENE data with a pretrained geneformer model</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Model Finetuning</p></td>
<td><p><a class="reference internal" href="notebooks/bionemo-finetuning-overview.html"><span class="doc std std-doc">Overview of Finetuning pre-trained models in BioNeMo</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Encoder Finetuning</p></td>
<td><p><a class="reference internal" href="notebooks/encoder-finetuning-notebook-fw.html"><span class="doc std std-doc">Encoder Fine-tuning in BioNeMo: MegaMolBART</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Downstream Tasks</p></td>
<td><p><a class="reference internal" href="notebooks/retrosynthesis-notebook.html"><span class="doc std std-doc">Training a Retrosynthesis Model using USPTO50 Dataset</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Downstream Tasks</p></td>
<td><p><a class="reference internal" href="notebooks/physchem-notebook-fw.html"><span class="doc std std-doc">Fine-tuning MegaMolBART for Solubility Prediction</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Custom Datasets</p></td>
<td><p><a class="reference internal" href="notebooks/custom-dataset-preprocessing-fw.html"><span class="doc std std-doc">Adding the OAS Dataset: Downloading and Preprocessing</span></a></p></td>
</tr>
<tr class="row-odd"><td><p>Custom Datasets</p></td>
<td><p><a class="reference internal" href="notebooks/custom-dataset-class-fw.html"><span class="doc std std-doc">Adding the OAS Dataset: Modifying the Dataset Class</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Custom DataLoaders</p></td>
<td><p><a class="reference internal" href="notebooks/custom-dataset-dataloader.html"><span class="doc std std-doc">Creating a Custom Dataloader</span></a></p></td>
</tr>
</tbody>
</table>
</div>
</div>

<div class="section">
   
</div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By NVIDIA<br/>
  
      &copy; Copyright 2023-2024 NVIDIA CORPORATION &amp; AFFILIATES. All rights reserved..<br/>
    Last updated on Oct 17, 2024.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>