
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction &#8212; NVIDIA BioNeMo Framework</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="https://js.hcaptcha.com/1/api.js"></script>
    <script src="//assets.adobedtm.com/5d4962a43b79/c1061d2c5e7b/launch-191c2462b890.min.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script src="../_static/design-tabs.js"></script>
    <link rel="canonical" href="https://docs.nvidia.com/bionemo-framework/0.4.0/notebooks/dnabert_pretrain_finetune.html" />
    <link rel="shortcut icon" href="../_static/nvidia-logo-vert-rgb-blk-for-screen.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Frequently Asked Questions (FAQ)" href="../faq-fw.html" />
    <link rel="prev" title="Generating and visualizing embeddings with DNABert" href="dnabert_inference.html" /> 
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
      
<style type="text/css">
  ul.ablog-archive {
    list-style: none;
    overflow: auto;
    margin-left: 0px;
  }
  ul.ablog-archive li {
    float: left;
    margin-right: 5px;
    font-size: 80%;
  }
  ul.postlist a {
    font-style: italic;
  }
  ul.postlist-style-disc {
    list-style-type: disc;
  }
  ul.postlist-style-none {
    list-style-type: none;
  }
  ul.postlist-style-circle {
    list-style-type: circle;
  }
</style>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/nvidia-logo-horiz-rgb-blk-for-screen.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">NVIDIA BioNeMo Framework</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GETTING STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../index.html">
   What is BioNeMo?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pre-reqs.html">
   Hardware and Software Prerequisites
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../access-startup.html">
   Access and Startup
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../initialization-guide.html">
   Initialization Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../tutorial-list.html">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BioNeMo CORE
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../bionemo-fw-for-model-training-fw.html">
   Fundamentals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../data-module-fw.html">
   Data Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../dwnstr-task-validation.html">
   Validation with a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../inference-triton-fw.html">
   Inference with Nvidia Triton Server
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../deep-dive-esm1-pytriton-inference.html">
   Example of PyTriton Inference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BEST PRACTICES
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../hyperparameters-fw.html">
   Hyperparameter Usage and Tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../parallelism-fw.html">
   Training Parallelism
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  MODELS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../models/diffdock.html">
   DiffDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/dnabert.html">
   DNABERT
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/dsmbind.html">
   DSMBind
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/equidock.html">
   EquiDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/esm1-nv.html">
   ESM-1nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/esm2-nv.html">
   ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/geneformer.html">
   Geneformer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/megamolbart.html">
   MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/molmim.html">
   MolMIM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/openfold.html">
   OpenFold
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/prott5nv.html">
   Prott5nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/model-benchmarks.html">
   Model Benchmarks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  DATASETS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/uniprot.html">
   UniProt Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/CELLxGENE.html">
   CELLxGENE
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/zinc15.html">
   ZINC-15 Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/pdb.html">
   The Protein Data Bank (PDB)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/GRCh38.p13.html">
   Human Reference Genome Version GRCh38.p13
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/openproteinset.html">
   OpenProteinSet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/moleculenet-physchem.html">
   MoleculeNet Physical Chemistry Datasets - Lipophilicity, FreeSolv, ESOL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  TUTORIALS
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing-bcp-training-diffdock.html">
   DiffDock: Preparing Workspace and Data for Pre-training
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2nv-mutant-design.html">
   Zero-Shot Protein Design Using ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MMB_GenerativeAI_Inference_with_examples.html">
   BioNeMo - MegaMolBART Inferencing for Generative Chemistry
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MolMIM_GenerativeAI_local_inference_with_examples.html">
   MolMIM Inferencing for Generative Chemistry and Downstream Prediction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ZINC15-data-preprocessing.html">
   ZINC Training Dataset Setup for small molecule language models
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bionemo-finetuning-overview.html">
   Finetune pre-trained models in BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cma_es_guided_molecular_optimization_molmim.html">
   MolMIM Property Guided Molecular Optimization Using CMA-ES
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-class-fw.html">
   Adding the OAS Dataset: Modifying the Dataset Class
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-dataloader.html">
   Adding the OAS Dataset: Customizing Dataset Object and Dataloader Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-preprocessing-fw.html">
   Adding the OAS Dataset: Downloading and Preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="encoder-finetuning-notebook-fw.html">
   Encoder Fine-tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_diffdock.html">
   DiffDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_equidock.html">
   EquiDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_esm1nv.html">
   ESM1nv Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_esm2nv.html">
   ESM-2nv: Data Preprocessing and Model Training Using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_mmb.html">
   MegaMolBART Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_molmim.html">
   Train MolMIM from scratch on your own data using the BioNeMo Framework
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="physchem-notebook-fw.html">
   Finetuning LLM in BioNeMo for a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="protein-esm2nv-clustering.html">
   Generating Embeddings for Protein Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="retrosynthesis-notebook.html">
   Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_FLIP_finetuning.html">
   Fine-Tune ESM-2nv on FLIP Data for Sequence-Level Classification, Regression, Token-Level Classification, and with LoRA Adapters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_oas_inferencing.html">
   Performing inference on OAS sequences with ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_paratope_finetuning.html">
   Pretrain from Scratch, Continue Training from an Existing Checkpoint, and Fine-tune ESM-2nv on Custom Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dnabert_inference.html">
   Generating and visualizing embeddings with DNABert
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  APPENDIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../faq-fw.html">
   Frequently Asked Questions (FAQ)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../bibliography-fw.html">
   Bibliography
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../releasenotes-fw.html">
   Release Notes
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#demo-objectives">
   Demo Objectives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#import-and-install-all-required-packages">
   Import and install all required packages
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#home-directory">
   Home Directory
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-model-checkpoints">
   Download Model Checkpoints
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-small-note-about-config-files-and-model-options">
     A Small Note About Config Files and Model Options
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocessing-data">
   1. Preprocessing data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pretrain-dnabert">
   2. Pretrain DNABert
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#further-pre-training-from-a-checkpoint">
   Further pre-training from a checkpoint
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fine-tuning-dnabert-splice-site-prediction-task">
   3. Fine-tuning DNABert: Splice site prediction task
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#splice-site-prediction-task">
   Splice-site prediction task
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-format-for-splice-site-prediction-task">
   Data Format for Splice Site Prediction Task
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocessors-overview">
   Preprocessors Overview
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-splice-site-prediction-task">
     Running splice site prediction task
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#demo-objectives">
   Demo Objectives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#import-and-install-all-required-packages">
   Import and install all required packages
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#home-directory">
   Home Directory
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-model-checkpoints">
   Download Model Checkpoints
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#a-small-note-about-config-files-and-model-options">
     A Small Note About Config Files and Model Options
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocessing-data">
   1. Preprocessing data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pretrain-dnabert">
   2. Pretrain DNABert
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#further-pre-training-from-a-checkpoint">
   Further pre-training from a checkpoint
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#fine-tuning-dnabert-splice-site-prediction-task">
   3. Fine-tuning DNABert: Splice site prediction task
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#splice-site-prediction-task">
   Splice-site prediction task
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#data-format-for-splice-site-prediction-task">
   Data Format for Splice Site Prediction Task
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocessors-overview">
   Preprocessors Overview
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-splice-site-prediction-task">
     Running splice site prediction task
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                 <div class="section" id="pretrain-fine-tune-and-perform-inference-with-dnabert-for-splice-site-prediction">
<h1>Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction<a class="headerlink" href="#pretrain-fine-tune-and-perform-inference-with-dnabert-for-splice-site-prediction" title="Permalink to this headline">#</a></h1>
<div class="alert alert-block alert-info"> <b>NOTE</b> This notebook was tested on a single RT5880 Ada Generation GPU using BioNeMo Framework v1.8 with an expected runtime of approximately <20 min.</div><div class="section" id="demo-objectives">
<h2>Demo Objectives<a class="headerlink" href="#demo-objectives" title="Permalink to this headline">#</a></h2>
<ol class="arabic simple">
<li><p><strong>Preprocess Data</strong></p>
<ul class="simple">
<li><p><strong>Objective</strong>: Download and prepare genomic data (FASTA and GFF3 files) for training and evaluation.</p></li>
<li><p><strong>Steps</strong>:</p>
<ul>
<li><p>Download and process FASTA files.</p></li>
<li><p>Extract splice site information from GFF3 files.</p></li>
<li><p>Organize data into train, validation, and test sets.</p></li>
</ul>
</li>
</ul>
</li>
<li><p><strong>Pretrain DNABERT</strong></p>
<ul class="simple">
<li><p><strong>Objective</strong>: Pretrain the DNABERT model on the processed genomic data.</p></li>
<li><p><strong>Steps</strong>:</p>
<ul>
<li><p>Load and preprocess genomic data.</p></li>
<li><p>Configure the DNABERT pretraining environment.</p></li>
<li><p>Execute pretraining and save the model checkpoint.</p></li>
<li><p>Implement further pretraining from a checkpoint.</p></li>
</ul>
</li>
</ul>
</li>
<li><p><strong>Fine-tune DNABERT for Splice Site Prediction</strong></p>
<ul class="simple">
<li><p><strong>Objective</strong>: Fine-tune DNABERT for accurate splice site prediction.</p></li>
<li><p><strong>Steps</strong>:</p>
<ul>
<li><p>Prepare train, validation, and test datasets for the splice site prediction task.</p></li>
<li><p>Load the pretrained model.</p></li>
<li><p>Set up the fine-tuning environment.</p></li>
<li><p>Train on splice site data and evaluate performance.</p></li>
</ul>
</li>
</ul>
</li>
</ol>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">#</a></h2>
<p>Ensure that you have read through the Getting Started section, can run the BioNeMo Framework docker container, and have configured the NGC Command Line Interface (CLI) within the container. It is assumed that this notebook is being executed from within the container.</p>
<div class="alert alert-block alert-info" style="padding: 15px; border: 1px solid #cfe2ff; background-color: #e7f3fe; border-radius: 5px;">
    <b>NOTE:</b> Some of the cells below can generate long text output. We're using:
    <pre style="background-color: #f8f9fa; border: 1px solid #dee2e6; padding: 10px; border-radius: 5px;">%%capture --no-display --no-stderr cell_output</pre>
    to suppress this output. Comment or delete this line in the cells below to restore full output.
</div></div>
<div class="section" id="import-and-install-all-required-packages">
<h2>Import and install all required packages<a class="headerlink" href="#import-and-install-all-required-packages" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">from</span> <span class="nn">bionemo.data.preprocess.dna.preprocess</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">GRCh38Ensembl99FastaResourcePreprocessor</span><span class="p">,</span>
    <span class="n">GRCh38Ensembl99GFF3ResourcePreprocessor</span>
<span class="p">)</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="home-directory">
<h2>Home Directory<a class="headerlink" href="#home-directory" title="Permalink to this headline">#</a></h2>
<p>Set the home directory as follows:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bionemo_home</span> <span class="o">=</span> <span class="s2">&quot;/workspace/bionemo&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;BIONEMO_HOME&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bionemo_home</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="download-model-checkpoints">
<h2>Download Model Checkpoints<a class="headerlink" href="#download-model-checkpoints" title="Permalink to this headline">#</a></h2>
<p>The following code will download the pretrained model dnabert-86M.nemo from the NGC registry.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define the NGC CLI API KEY and ORG for the model download</span>
<span class="c1"># If these variables are not already set in the container, uncomment below</span>
<span class="c1"># to define and set with your API KEY and ORG</span>

<span class="c1"># api_key = &lt;YOUR_API_KEY&gt;</span>
<span class="c1"># ngc_cli_org = &lt;YOUR_NGC_ORG&gt;</span>
<span class="c1"># Update the environment variables </span>
<span class="c1"># os.environ[&#39;NGC_CLI_API_KEY&#39;] = api_key</span>
<span class="c1"># os.environ[&#39;NGC_CLI_ORG&#39;] = ngc_cli_org</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Set variables and paths for model and checkpoint</span>
<span class="n">model_name</span> <span class="o">=</span> <span class="s2">&quot;dnabert&quot;</span> 
<span class="n">model_version</span> <span class="o">=</span> <span class="s2">&quot;dnabert-86M&quot;</span> 
<span class="n">actual_checkpoint_name</span> <span class="o">=</span> <span class="s2">&quot;dnabert-86M.nemo&quot;</span> 
<span class="n">model_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="s1">&#39;models&#39;</span><span class="p">)</span>
<span class="n">checkpoint_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_path</span><span class="p">,</span> <span class="n">actual_checkpoint_name</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;MODEL_PATH&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_path</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
if not os.path.exists(checkpoint_path):
    !cd /workspace/bionemo &amp;&amp; \
    python download_artifacts.py --model_dir models --models {model_name}
else:
    print(f&quot;Model {model_name} already exists at {model_path}.&quot;)
</pre></div>
</div>
</div>
</div>
<div class="section" id="a-small-note-about-config-files-and-model-options">
<h3>A Small Note About Config Files and Model Options<a class="headerlink" href="#a-small-note-about-config-files-and-model-options" title="Permalink to this headline">#</a></h3>
<p>When working with different variants of DNABERT, such as the regular version or the <code class="docutils literal notranslate"><span class="pre">xsmall</span></code> version, it’s important to ensure that the configuration files used for pretraining, fine-tuning, and downstream tasks match the model architecture. If you pretrain using the regular DNABERT, you must use the corresponding configuration file for fine-tuning and other tasks. Conversely, if you opt for the <code class="docutils literal notranslate"><span class="pre">xsmall</span></code> version of DNABERT, make sure to adjust your configuration files accordingly to maintain consistency. The configuration files for these models can be found in the <code class="docutils literal notranslate"><span class="pre">examples/dna/dnabert/conf/</span></code> directory. Mismatched configurations can lead to errors, particularly size mismatches in model layers, as the model architectures differ between variants. For the purposes of this tutorial, we are using the <code class="docutils literal notranslate"><span class="pre">dnabert_xsmall</span></code> config, which makes the tutorial easily executable on a single GPU, as the <code class="docutils literal notranslate"><span class="pre">dnabert_xsmall</span></code> model only has 8.1M parameters compared to the regular DNABert’s 86M.</p>
</div>
</div>
<div class="section" id="preprocessing-data">
<h2>1. Preprocessing data<a class="headerlink" href="#preprocessing-data" title="Permalink to this headline">#</a></h2>
<p>DNABERT model version dnabert-86M.nemo was pretrained on the GRCh38 human genome assembly downloaded from NCBI. From here, you can perform downstream tasks such as splice site prediction, which will be demonstrated later on in the tutorial.</p>
<p>This script prepares the data by running the necessary preprocessing steps before you begin training DNABERT on the GRCh38 human genome assembly. The preprocessing steps include downloading the genomic data, chunking the sequences, and organizing the data into structured FASTA/CSV files for easy access during training. The preprocessed data is stored in the specified directory, which will be used in subsequent training and downstream tasks. The script splits the preprocessed genome data into training, validation, and testing sets. The genome data is split by chromosome: chromosomes 1 through 19 are placed in the training set, chromosome 20 is the validation set, and chromosome 21 is the test set.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
processed_pretraining_data = os.path.join(bionemo_home, &#39;data&#39;, &#39;GRCh38.p13&#39;)
config_dir = os.path.join(bionemo_home, &#39;examples&#39;, &#39;dna&#39;, &#39;dnabert&#39;, &#39;conf&#39;)

# Run the preprocessing step
!cd {bionemo_home} &amp;&amp; python examples/dna/dnabert/pretrain.py \
  --config-path={config_dir} \
  --config-name=dnabert_xsmall \
  ++do_training=False \
  ++model.data.dataset_path={processed_pretraining_data}
</pre></div>
</div>
</div>
</div>
<p>Now, once we have obtained our preprocessed sequences, we can continue with the pretraining step.</p>
</div>
<div class="section" id="pretrain-dnabert">
<h2>2. Pretrain DNABert<a class="headerlink" href="#pretrain-dnabert" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
! cd {bionemo_home} &amp;&amp; python examples/dna/dnabert/pretrain.py \
  --config-path={config_dir} \
  --config-name=dnabert_xsmall \
  ++trainer.devices=1 \
  ++trainer.max_steps=1 \
  ++trainer.val_check_interval=1 \
  ++model.data.dataset_path={processed_pretraining_data} \
  ++model.data.dataset.train=chr\\[1..19\\].fna.gz.chunked.fa \
  ++model.data.dataset.val=chr20.fna.gz.chunked.fa \
  ++model.data.dataset.test=chr21.fna.gz.chunked.fa \
  ++exp_manager.create_wandb_logger=false
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="further-pre-training-from-a-checkpoint">
<h2>Further pre-training from a checkpoint<a class="headerlink" href="#further-pre-training-from-a-checkpoint" title="Permalink to this headline">#</a></h2>
<p>In this section, we will explore how to resume training from a pre-existing checkpoint. The script below demonstrates how to initiate further training using a specific checkpoint, ensuring that your model benefits from all previous training efforts. To continue training the model, you must increase the <code class="docutils literal notranslate"><span class="pre">max_steps</span></code> value, as it retains the previous pretraining metadata and will stop if <code class="docutils literal notranslate"><span class="pre">max_steps</span></code> is already reached.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
# Run the pretraining script with these paths
! cd {bionemo_home} &amp;&amp; python examples/dna/dnabert/pretrain.py \
  --config-path={config_dir} \
  --config-name=dnabert_xsmall \
  ++do_training=True \
  ++trainer.devices=1 \
  ++trainer.max_steps=2 \
  ++trainer.val_check_interval=1 \
  ++model.data.dataset_path={dataset_dir} \
  ++exp_manager.create_wandb_logger=false \
  ++exp_manager.resume_if_exists=True
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="fine-tuning-dnabert-splice-site-prediction-task">
<h2>3. Fine-tuning DNABert: Splice site prediction task<a class="headerlink" href="#fine-tuning-dnabert-splice-site-prediction-task" title="Permalink to this headline">#</a></h2>
</div>
<div class="section" id="splice-site-prediction-task">
<h2>Splice-site prediction task<a class="headerlink" href="#splice-site-prediction-task" title="Permalink to this headline">#</a></h2>
<p>In this task, we will utilize the DNABERT model for predicting splice sites within the human genome. Splice sites are critical regions in the DNA sequence where introns are removed, and exons are joined together during gene expression. We will be working with a dataset specifically prepared using the GRCh38Ensembl99FastaResourcePreprocessor, which provides the necessary .fa.gz files containing raw DNA sequences. These files have been preprocessed to create training, validation, and test datasets tailored for the splice site prediction task.</p>
</div>
<div class="section" id="data-format-for-splice-site-prediction-task">
<h2>Data Format for Splice Site Prediction Task<a class="headerlink" href="#data-format-for-splice-site-prediction-task" title="Permalink to this headline">#</a></h2>
<p>For the splice site prediction task, the data is formatted in CSV files where each row corresponds to a genomic sequence centered around a potential splice site. The key columns in these CSV files include the following:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">id</span></code></strong>: This column represents a unique identifier for each row in the dataset.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">coord</span></code></strong>: This column specifies the exact coordinate or position on the chromosome where the splice site is located. The coordinate is typically the center of the sequence window that is being analyzed.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">kind</span></code></strong>: This column indicates the type of splice site, where:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">0</span></code> = Donor site</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">1</span></code> = Acceptor site</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">2</span></code> = Negative example (non-splice site region)</p></li>
</ul>
</li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">transcript</span></code></strong>: This column contains detailed transcript information, including the identifier for the transcript and potentially the chromosome number.</p></li>
</ul>
<p>The sequences themselves are extracted from FASTA files based on the coordinates provided in the CSV and are processed into k-mer representations suitable for input into DNABERT. Each sequence is typically centered on a candidate splice site, allowing the model to learn the characteristics that distinguish true splice sites from non-functional regions.</p>
<p>After running the preprocessor, the preprocessed data files are moved to their respective directories (train, val, test) to ensure they are correctly organized for the subsequent steps in the pipeline. This ensures that the fine-tuning process have access to the properly formatted and placed data.</p>
</div>
<div class="section" id="preprocessors-overview">
<h2>Preprocessors Overview<a class="headerlink" href="#preprocessors-overview" title="Permalink to this headline">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">preprocess.py</span></code> file contains multiple classes that handle different types of genomic data, depending on the task and dataset. These preprocessors are designed to handle the retrieval, preparation, and formatting of data required for various types of genomic analyses:</p>
<ol class="arabic simple">
<li><p><strong>GRCh38p13_ResourcePreprocessor</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This preprocessor is tailored for the GRCh38.p13 human genome assembly, specifically designed to download all primary chromosomes from this version. It handles the preparation of a set of files, ensuring that each chromosome’s sequence is correctly retrieved and stored for further genomic analyses.</p></li>
</ul>
</li>
<li><p><strong>Hg38chromResourcePreprocessor</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This preprocessor is designed to download the hg38 chromosome sequences from the UCSC Genome Browser. It is closely tied to specific datasets and provides a structured way to obtain and prepare these sequences for downstream analyses.</p></li>
</ul>
</li>
<li><p><strong>GRCh38Ensembl99FastaResourcePreprocessor</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This preprocessor is intended for downloading and preparing the FASTA files for the GRCh38 Ensembl release 99. It focuses on retrieving the chromosome sequences in the <code class="docutils literal notranslate"><span class="pre">.fa.gz</span></code> format, ensuring they are correctly formatted for tasks like sequence analysis or prediction.</p></li>
</ul>
</li>
<li><p><strong>GRCh38Ensembl99GFF3ResourcePreprocessor</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This preprocessor is used for downloading GFF3 files from Ensembl release 99, which contain annotations and features required for splice site prediction tasks, as utilized in the DNABERT publication. It ensures that the correct genomic annotations are available for these analyses.</p></li>
</ul>
</li>
<li><p><strong>DNABERTPreprocessorDataClass</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This class provides a structured way to initialize and configure the DNABERT preprocessing pipeline. It includes necessary configurations like genome directory paths, tokenizer models, and dataset configuration, essential for setting up the DNABERT model’s preprocessing phase.</p></li>
</ul>
</li>
<li><p><strong>CorePromoterResourcePreparer</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This preprocessor focuses on downloading the necessary files for core promoter prediction. It is tightly coupled with specific datasets, such as those from the HPDnew database, to ensure the correct files are prepared for analyzing promoter regions in the genome.</p></li>
</ul>
</li>
<li><p><strong>BasenjiDatasetPreprocessor</strong>:</p>
<ul class="simple">
<li><p><strong>Description</strong>: This preprocessor is responsible for downloading the Basenji2 dataset in its original TFRecord format. It then converts the dataset to WebDataset format and reorganizes metadata. This preprocessor is essential for tasks involving the Basenji2 dataset, particularly in genomic prediction models.</p></li>
</ul>
</li>
</ol>
<p>Each of these preprocessors is designed for a specific type of genomic data or task, ensuring that the data is correctly retrieved, prepared, and formatted for downstream analyses.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span> --no-display --no-stderr cell_output
<span class="c1"># Setting paths</span>
<span class="n">fasta_directory</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="s1">&#39;examples/dna/data/splice-site-prediction/GRCh38.ensembl.99&#39;</span><span class="p">)</span>

<span class="c1"># Instantiating the preprocessor</span>
<span class="n">preprocessor</span> <span class="o">=</span> <span class="n">GRCh38Ensembl99FastaResourcePreprocessor</span><span class="p">(</span><span class="n">root_directory</span><span class="o">=</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="n">dest_directory</span><span class="o">=</span><span class="n">fasta_directory</span><span class="p">)</span>

<span class="c1"># Running the preprocessor to download and prepare the dataset</span>
<span class="n">downloaded_files</span> <span class="o">=</span> <span class="n">preprocessor</span><span class="o">.</span><span class="n">prepare</span><span class="p">()</span>

<span class="c1"># Output paths for reference</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Downloaded Files:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">downloaded_files</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now, we need to use the GRCh38Ensembl99GFF3ResourcePreprocessor to generate the train, test, and val .csv files that follow the aforementioned format for the task.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span> --no-display --no-stderr cell_output
<span class="n">finetuning_dataset_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="s1">&#39;examples/dna/data/splice-site-prediction/finetuning_data&#39;</span><span class="p">)</span>

<span class="c1"># Ensuring the target directory exists</span>
<span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">finetuning_dataset_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Instantiating the GFF3 preprocessor</span>
<span class="n">gff3_preprocessor</span> <span class="o">=</span> <span class="n">GRCh38Ensembl99GFF3ResourcePreprocessor</span><span class="p">(</span>
    <span class="n">root_directory</span><span class="o">=</span><span class="n">bionemo_home</span><span class="p">,</span> 
    <span class="n">dest_directory</span><span class="o">=</span><span class="n">finetuning_dataset_dir</span>  
<span class="p">)</span>

<span class="c1"># Run the preprocessor to download and prepare the dataset (train, val, test CSV files)</span>
<span class="n">csv_files</span> <span class="o">=</span> <span class="n">gff3_preprocessor</span><span class="o">.</span><span class="n">prepare</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Generated CSV Files:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">csv_files</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="running-splice-site-prediction-task">
<h3>Running splice site prediction task<a class="headerlink" href="#running-splice-site-prediction-task" title="Permalink to this headline">#</a></h3>
<p>Below, we set up the file paths for the command line arguments. We set <code class="docutils literal notranslate"><span class="pre">do_prediction=True</span></code> in the command line arguments to get a .txt file of predictions that can be used for evaluation purposes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">train_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">finetuning_dataset_dir</span><span class="p">,</span> <span class="s1">&#39;train.csv&#39;</span><span class="p">)</span>
<span class="n">val_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">finetuning_dataset_dir</span><span class="p">,</span> <span class="s1">&#39;val.csv&#39;</span><span class="p">)</span>
<span class="n">test_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">finetuning_dataset_dir</span><span class="p">,</span> <span class="s1">&#39;test.csv&#39;</span><span class="p">)</span>

<span class="c1"># Print to verify the paths</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Config Directory:&quot;</span><span class="p">,</span> <span class="n">config_dir</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Dataset Directory:&quot;</span><span class="p">,</span> <span class="n">finetuning_dataset_dir</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Pretrained Model Path:&quot;</span><span class="p">,</span> <span class="n">checkpoint_path</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Fasta Directory:&quot;</span><span class="p">,</span> <span class="n">fasta_directory</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Config Directory: /workspace/bionemo/examples/dna/dnabert/conf
Dataset Directory: /workspace/bionemo/examples/dna/data/splice-site-prediction/finetuning_data
Pretrained Model Path: /workspace/bionemo/models/dnabert-86M.nemo
Fasta Directory: /workspace/bionemo/examples/dna/data/splice-site-prediction/GRCh38.ensembl.99
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
! cd {bionemo_home} &amp;&amp; python examples/dna/dnabert/downstream_splice_site.py \
  --config-path={config_dir} \
  --config-name=dnabert_config_splice_site \
  ++do_training=True \
  ++do_prediction=True \
  ++trainer.devices=1 \
  ++trainer.max_steps=3 \
  ++trainer.max_epochs=1 \
  ++trainer.val_check_interval=1 \
  ++model.encoder_frozen=False \
  ++model.data.dataset_path={finetuning_dataset_dir} \
  ++model.data.train_file={train_file} \
  ++model.data.val_file={val_file} \
  ++model.data.predict_file={test_file} \
  ++model.restore_encoder_path={checkpoint_path} \
  ++exp_manager.create_wandb_logger=false \
  ++exp_manager.resume_if_exists=False 
</pre></div>
</div>
</div>
</div>
<p>In this demo, we demonstrated how to fine-tune the DNABERT model for the task of splice site prediction. We covered the steps of data preprocessing, model pretraining, and fine-tuning on splice site data. Finally, we generated predictions on a test dataset, which are saved in <code class="docutils literal notranslate"><span class="pre">results/nemo_experiments/dnabert-splicesite/dnabert-splicesite</span></code>. These predictions can be further analyzed or used as a foundation for additional evaluations, depending on the specific needs of your project.</p>
</div>
</div>
</div>

<div class="section">
   
</div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="dnabert_inference.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Generating and visualizing embeddings with DNABert</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../faq-fw.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Frequently Asked Questions (FAQ)</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By NVIDIA<br/>
  
      &copy; Copyright 2023-2024 NVIDIA CORPORATION &amp; AFFILIATES. All rights reserved..<br/>
    Last updated on Nov 20, 2024.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>