
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ESM-2nv: Data Preprocessing and Model Training Using BioNeMo &#8212; NVIDIA BioNeMo Framework</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="https://js.hcaptcha.com/1/api.js"></script>
    <script src="//assets.adobedtm.com/5d4962a43b79/c1061d2c5e7b/launch-191c2462b890.min.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script src="../_static/design-tabs.js"></script>
    <link rel="canonical" href="https://docs.nvidia.com/bionemo-framework/0.4.0/notebooks/model_training_esm2nv.html" />
    <link rel="shortcut icon" href="../_static/nvidia-logo-vert-rgb-blk-for-screen.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="MegaMolBART Model Training using BioNeMo" href="model_training_mmb.html" />
    <link rel="prev" title="ESM1nv Model Training using BioNeMo" href="model_training_esm1nv.html" /> 
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
      
<style type="text/css">
  ul.ablog-archive {
    list-style: none;
    overflow: auto;
    margin-left: 0px;
  }
  ul.ablog-archive li {
    float: left;
    margin-right: 5px;
    font-size: 80%;
  }
  ul.postlist a {
    font-style: italic;
  }
  ul.postlist-style-disc {
    list-style-type: disc;
  }
  ul.postlist-style-none {
    list-style-type: none;
  }
  ul.postlist-style-circle {
    list-style-type: circle;
  }
</style>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/nvidia-logo-horiz-rgb-blk-for-screen.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">NVIDIA BioNeMo Framework</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GETTING STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../index.html">
   What is BioNeMo?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pre-reqs.html">
   Hardware and Software Prerequisites
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../access-startup.html">
   Access and Startup
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../initialization-guide.html">
   Initialization Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../tutorial-list.html">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BioNeMo CORE
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../bionemo-fw-for-model-training-fw.html">
   Fundamentals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../data-module-fw.html">
   Data Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../dwnstr-task-validation.html">
   Validation with a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../inference-triton-fw.html">
   Inference with Nvidia Triton Server
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../deep-dive-esm1-pytriton-inference.html">
   Example of PyTriton Inference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BEST PRACTICES
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../hyperparameters-fw.html">
   Hyperparameter Usage and Tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../parallelism-fw.html">
   Training Parallelism
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  MODELS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../models/diffdock.html">
   DiffDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/dnabert.html">
   DNABERT
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/dsmbind.html">
   Model Overview
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/equidock.html">
   EquiDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/esm1-nv.html">
   ESM-1nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/esm2-nv.html">
   ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/geneformer.html">
   Geneformer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/megamolbart.html">
   MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/molmim.html">
   MolMIM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/openfold.html">
   OpenFold
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/prott5nv.html">
   Prott5nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/model-benchmarks.html">
   Model Benchmarks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  DATASETS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/uniprot.html">
   UniProt Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/CELLxGENE.html">
   CELLxGENE
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/zinc15.html">
   ZINC-15 Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/pdb.html">
   The Protein Data Bank (PDB)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/GRCh38.p13.html">
   Human Reference Genome Version GRCh38.p13
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/openproteinset.html">
   OpenProteinSet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/moleculenet-physchem.html">
   MoleculeNet Physical Chemistry Datasets - Lipophilicity, FreeSolv, ESOL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  TUTORIALS
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing-bcp-training-diffdock.html">
   DiffDock: Preparing Workspace and Data for Pre-training
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2nv-mutant-design.html">
   Zero-Shot Protein Design Using ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MMB_GenerativeAI_Inference_with_examples.html">
   BioNeMo - MegaMolBART Inferencing for Generative Chemistry
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MolMIM_GenerativeAI_local_inference_with_examples.html">
   MolMIM Inferencing for Generative Chemistry and Downstream Prediction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ZINC15-data-preprocessing.html">
   ZINC Training Dataset Setup for small molecule language models
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bionemo-finetuning-overview.html">
   Finetune pre-trained models in BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cma_es_guided_molecular_optimization_molmim.html">
   MolMIM Property Guided Molecular Optimization Using CMA-ES
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-class-fw.html">
   Adding the OAS Dataset: Modifying the Dataset Class
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-dataloader.html">
   Adding the OAS Dataset: Customizing Dataset Object and Dataloader Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-preprocessing-fw.html">
   Adding the OAS Dataset: Downloading and Preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="encoder-finetuning-notebook-fw.html">
   Encoder Fine-tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_diffdock.html">
   DiffDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_equidock.html">
   EquiDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_esm1nv.html">
   ESM1nv Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   ESM-2nv: Data Preprocessing and Model Training Using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_mmb.html">
   MegaMolBART Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_molmim.html">
   Train MolMIM from scratch on your own data using the BioNeMo Framework
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="physchem-notebook-fw.html">
   Finetuning LLM in BioNeMo for a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="protein-esm2nv-clustering.html">
   Generating Embeddings for Protein Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="retrosynthesis-notebook.html">
   Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_FLIP_finetuning.html">
   Fine-Tune ESM-2nv on FLIP Data for Sequence-Level Classification, Regression, Token-Level Classification, and with LoRA Adapters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_oas_inferencing.html">
   Performing inference on OAS sequences with ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_paratope_finetuning.html">
   Pretrain from Scratch, Continue Training from an Existing Checkpoint, and Fine-tune ESM-2nv on Custom Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dnabert_inference.html">
   Generating and visualizing embeddings with DNABert
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dnabert_pretrain_finetune.html">
   Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  APPENDIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../faq-fw.html">
   Frequently Asked Questions (FAQ)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../bibliography-fw.html">
   Bibliography
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../releasenotes-fw.html">
   Release Notes
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#demo-objectives">
   Demo Objectives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-esm-2nv-model">
   Overview - ESM-2nv Model
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#import-and-install-all-required-packages">
   Import and install all required packages
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#home-directory">
   Home Directory
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-model-checkpoints">
   Download Model Checkpoints
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocess-data-for-esm-2nv">
   Preprocess Data for ESM-2nv
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#command-line-and-yaml-configuration-for-pretrain-py">
     Command Line and YAML Configuration for
     <code class="docutils literal notranslate">
      <span class="pre">
       pretrain.py
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pretrain-from-scratch">
   Pretrain from Scratch
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#continue-training-from-an-existing-checkpoint">
   Continue Training from an Existing Checkpoint
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>ESM-2nv: Data Preprocessing and Model Training Using BioNeMo</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#demo-objectives">
   Demo Objectives
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview-esm-2nv-model">
   Overview - ESM-2nv Model
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup">
   Setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#import-and-install-all-required-packages">
   Import and install all required packages
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#home-directory">
   Home Directory
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#download-model-checkpoints">
   Download Model Checkpoints
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#preprocess-data-for-esm-2nv">
   Preprocess Data for ESM-2nv
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#command-line-and-yaml-configuration-for-pretrain-py">
     Command Line and YAML Configuration for
     <code class="docutils literal notranslate">
      <span class="pre">
       pretrain.py
      </span>
     </code>
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pretrain-from-scratch">
   Pretrain from Scratch
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#continue-training-from-an-existing-checkpoint">
   Continue Training from an Existing Checkpoint
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                 <div class="tex2jax_ignore mathjax_ignore section" id="esm-2nv-data-preprocessing-and-model-training-using-bionemo">
<h1>ESM-2nv: Data Preprocessing and Model Training Using BioNeMo<a class="headerlink" href="#esm-2nv-data-preprocessing-and-model-training-using-bionemo" title="Permalink to this headline">#</a></h1>
<div class="alert alert-block alert-info">
  <b>NOTE:</b> This notebook has been tested on both an A1000 GPU and an A100, and is compatible with BioNeMo Framework v1.6, v1.7 and v1.8. The expected runtime is less than 1 hour on the A1000 and ~3 minutes on the A100.
</div>
<div class="section" id="demo-objectives">
<h2>Demo Objectives<a class="headerlink" href="#demo-objectives" title="Permalink to this headline">#</a></h2>
<p>The purpose of this tutorial is to provide an example use case for training a BioNeMo large language model using the BioNeMo framework. In this tutorial, you will gain experience in:</p>
<ol class="arabic simple">
<li><p>Preprocessing the UniRef50 and UniRef90 data for ESM-2nv.</p></li>
<li><p>Pretraining and continuing training from a checkpoint for ESM-2nv.</p></li>
<li><p>Performing inference with ESM-2nv.</p></li>
</ol>
</div>
<div class="section" id="overview-esm-2nv-model">
<h2>Overview - ESM-2nv Model<a class="headerlink" href="#overview-esm-2nv-model" title="Permalink to this headline">#</a></h2>
<p><strong>ESM-2nv</strong> is based on the public ESM-2 model, which is a BERT architecture trained on millions of protein sequences from the UniProt database. ESM-2nv learns the patterns and dependencies between amino acids that ultimately give rise to a protein’s 2D structure. These can include properties such as alpha helix or beta sheet, as well as cellular location, thermostability, solubility, and other protein properties. For more information, check the <a class="reference internal" href="../models/esm2-nv.html"><span class="doc std std-doc">ESM-2nv model card</span></a></p>
<p>This ESM-2nv model training example walkthrough will show how to utilize compute resources, download and preprocess datasets, and perform model training on single and multiple nodes.</p>
<p>The model was trained on UniRef50 and UniRef90 protein sequences, truncated to a maximum length of 1022 amino acids.</p>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">#</a></h2>
<p>Ensure that you have read through the <a class="reference internal" href="../index.html"><span class="doc std std-doc">Getting Started</span></a> section, can run the BioNeMo Framework Docker container, and have configured the NGC Command Line Interface (CLI) within the container. It is assumed that this notebook is being executed from within the container. Additionally, this tutorial depends on the <a class="reference internal" href="../models/esm2-nv.html"><span class="doc std std-doc">ESM-2nv</span></a> model.</p>
<div class="alert alert-block alert-info"> <b>NOTE</b> Some of the cells below generate long text output.  We're using <pre>%%capture --no-display --no-stderr cell_output</pre> to suppress this output.  Comment or delete this line in the cells below to restore full output.</div>
</div>
<div class="section" id="import-and-install-all-required-packages">
<h2>Import and install all required packages<a class="headerlink" href="#import-and-install-all-required-packages" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="home-directory">
<h2>Home Directory<a class="headerlink" href="#home-directory" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">bionemo_home</span> <span class="o">=</span> <span class="s2">&quot;/workspace/bionemo&quot;</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;BIONEMO_HOME&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bionemo_home</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="download-model-checkpoints">
<h2>Download Model Checkpoints<a class="headerlink" href="#download-model-checkpoints" title="Permalink to this headline">#</a></h2>
<p>The following code will download the pretrained model <code class="docutils literal notranslate"><span class="pre">esm2nv_650M_converted.nemo</span></code> from the NGC registry.</p>
<p>In BioNeMo FW, there are numerous ESM models available, including ESM-1nv, ESM-2nv 8M with randomly initialized weights, ESM-2nv fine-tuned for secondary structure downstream prediction tasks with LoRA, ESM-2nv 650M, and ESM-2nv 3B. We also have a configuration file for training ESM-2nv 15B available at <code class="docutils literal notranslate"><span class="pre">examples/protein/esm2nv/conf/pretrain_esm2_15B.yaml</span></code> if needed.</p>
<p>For demo purposes, we have chosen to showcase the ESM-2nv 650M model. For more details on the <a class="reference external" href="https://docs.nvidia.com/bionemo-framework/latest/models/esm1-nv.html">ESM-1nv</a> or <a class="reference external" href="https://docs.nvidia.com/bionemo-framework/latest/models/esm2-nv.html">ESM-2nv</a>, consult the corresponding model cards. To find the model names and checkpoint names, please see the <code class="docutils literal notranslate"><span class="pre">artifacts_paths.yaml</span></code> file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define the NGC CLI API KEY and ORG for the model download</span>
<span class="c1"># If these variables are not already set in the container, uncomment below</span>
<span class="c1"># to define and set with your API KEY and ORG</span>
<span class="c1">#api_key = &lt;your_api_key&gt;</span>
<span class="c1">#ngc_cli_org = &lt;ngc_cli_org&gt;</span>
<span class="c1"># Update the environment variable</span>
<span class="c1">#os.environ[&#39;NGC_CLI_API_KEY&#39;] = api_key</span>
<span class="c1">#os.environ[&#39;NGC_CLI_ORG&#39;] = ngc_cli_org</span>

<span class="c1"># Set variables and paths for model and checkpoint</span>
<span class="n">model_name</span> <span class="o">=</span> <span class="s2">&quot;esm2nv&quot;</span> <span class="c1"># change to esm1nv for ESM1</span>
<span class="n">model_version</span> <span class="o">=</span> <span class="s2">&quot;esm2nv_650m&quot;</span> <span class="c1"># change to esm1nv for ESM1</span>
<span class="n">actual_checkpoint_name</span> <span class="o">=</span> <span class="s2">&quot;esm2nv_650M_converted.nemo&quot;</span> <span class="c1">#  change to esm1nv.nemo for ESM1</span>
<span class="n">model_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="s1">&#39;models&#39;</span><span class="p">)</span>
<span class="n">checkpoint_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">model_path</span><span class="p">,</span> <span class="n">actual_checkpoint_name</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;MODEL_PATH&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model_path</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
if not os.path.exists(checkpoint_path):
    !cd /workspace/bionemo &amp;&amp; \
    python download_artifacts.py --model_dir models --models {model_version}
else:
    print(f&quot;Model {model_version} already exists at {model_path}.&quot;)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="preprocess-data-for-esm-2nv">
<h2>Preprocess Data for ESM-2nv<a class="headerlink" href="#preprocess-data-for-esm-2nv" title="Permalink to this headline">#</a></h2>
<p>To briefly showcase the model training capabilities of the BioNeMo Framework, we will use the UniRef50 and UniRef90 datasets to provide a diverse yet non-redundant set of protein sequences. By using both, the model can learn from a wide range of sequence variants while avoiding redundancy. This helps in capturing diverse features and patterns that are relevant for protein function and structure prediction, while also preventing overfitting and improving generalization. For demo purposes, a portion of the sample datasets is located in <code class="docutils literal notranslate"><span class="pre">${bionemo_home}/examples/tests/test_data/uniref202104_esm2_qc</span></code>.</p>
<p>The data is stored in a zip file, so run the following command to extract the raw FASTA files and a cluster mapping file:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
# Define the path to the extracted directory
datapath_dir = os.path.join(bionemo_home, &#39;examples/tests/test_data/protein/uniref50_90&#39;)

# Define the path to the zip file
zip_file = f&quot;{datapath_dir}.zip&quot;

# Check if the directory already exists
if not os.path.exists(datapath_dir): 
    ! unzip {zip_file} -d {bionemo_home}/examples/tests/test_data/
else:
    pass
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">mapping.tsv</span></code> file is used to associate protein sequences with their respective clusters. This helps to reduce redundancy, organize data, and evaluate model performance by tracking sequence similarity and ensuring diverse training data.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">cluster_mapping.tsv</span></code> is used to associate protein sequences with their respective clusters, helping to reduce redundancy, organize data, and evaluate model performance by tracking sequence similarity and ensuring diverse training data.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">mapping.tsv</span></code> file is used to associate protein sequences with their respective clusters. This helps to reduce redundancy, organize data, and evaluate model performance by tracking sequence similarity and ensuring diverse training data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">train_uf50_fasta</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">datapath_dir</span><span class="si">}</span><span class="s1">/uniref50_train_filt.fasta&#39;</span><span class="p">)</span>
<span class="n">train_uf90_fasta</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">datapath_dir</span><span class="si">}</span><span class="s1">/ur90_ur50_sampler.fasta&#39;</span><span class="p">)</span>
<span class="n">train_cluster_mapping_tsv</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bionemo_home</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">datapath_dir</span><span class="si">}</span><span class="s1">/mapping.tsv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Using the unzipped contents of this file, we first create the preprocessed <code class="docutils literal notranslate"><span class="pre">/train</span></code>, <code class="docutils literal notranslate"><span class="pre">/val</span></code>, and <code class="docutils literal notranslate"><span class="pre">/test</span></code> folders, organizing protein sequences into batch CSV files. It is important to utilize both datasets if you plan to use ESM-2 as originally created. However, if you use your own data, as demonstrated in <a class="reference internal" href="esm2_paratope_finetuning.html"><span class="doc std std-doc">this notebook</span></a>, you can opt to use a single data source.</p>
<p>The same approach applies to the clustering mapping file. The <code class="docutils literal notranslate"><span class="pre">ESM2Preprocess</span></code> class can handle clustering indirectly as part of the dataset preparation process. It leverages UniRef50 to UniRef90 clustering mappings to organize protein sequences, ensuring that data is appropriately clustered for training and validation.</p>
<p>Please note that this script does not perform clustering itself but relies on pre-defined clustering mappings provided in a TSV file format to organize protein sequences. The expected format is a TSV file where the first column represents the cluster ID (FASTA header in UniRef50) and the second column lists the members separated by commas. The members correspond to entries in the UniRef90 FASTA file.</p>
<p>The preprocessing steps are:</p>
<ol class="arabic simple">
<li><p>Download the dataset from a specified URL or NGC registry.</p></li>
<li><p>Extract and decompress the downloaded data if necessary.</p></li>
<li><p>Index the FASTA file using <code class="docutils literal notranslate"><span class="pre">pyfastx</span></code> to facilitate data access.</p></li>
<li><p>Split the dataset into training, validation, and test sets.</p></li>
<li><p>Convert the FASTA sequences into CSV format, dividing them into multiple files if needed.</p></li>
<li><p>Generate additional files like memmaps or sorted FASTA files if required for specific use cases.</p></li>
</ol>
<p>For more details about the preprocessing steps, please consult the <code class="docutils literal notranslate"><span class="pre">.../bionemo/data/preprocess/protein/preprocess.py</span></code> file and the documentation found <span class="xref myst">here</span>.</p>
<p>To preprocess the data defined in the previous cell, use the <code class="docutils literal notranslate"><span class="pre">pretrain.py</span></code> script and set the <code class="docutils literal notranslate"><span class="pre">do_training</span></code> parameter to <code class="docutils literal notranslate"><span class="pre">False</span></code>, as shown below:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
! python examples/protein/esm2nv/pretrain.py \
    --config-path=conf \
    --config-name=pretrain_esm2_650M \
    ++do_training=False \
    ++do_preprocessing=True \
    ++model.data.val_size=500 \
    ++model.data.test_size=100 \
    ++model.data.train.uf50_datapath={train_uf50_fasta} \
    ++model.data.train.uf90_datapath={train_uf90_fasta} \
    ++model.data.train.cluster_mapping_tsv={train_cluster_mapping_tsv} \
    ++model.data.dataset_path={datapath_dir}
</pre></div>
</div>
</div>
</div>
<div class="section" id="command-line-and-yaml-configuration-for-pretrain-py">
<h3>Command Line and YAML Configuration for <code class="docutils literal notranslate"><span class="pre">pretrain.py</span></code><a class="headerlink" href="#command-line-and-yaml-configuration-for-pretrain-py" title="Permalink to this headline">#</a></h3>
<p>Parameters starting with <code class="docutils literal notranslate"><span class="pre">--</span></code> are passed as command line arguments to <code class="docutils literal notranslate"><span class="pre">pretrain.py</span></code>. Below are examples of such parameters:</p>
<ul>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">--config-path</span></code></strong> and <strong><code class="docutils literal notranslate"><span class="pre">--config-name</span></code></strong>:<br />
These specify the folder and the YAML file name for the configuration. The path is relative to <code class="docutils literal notranslate"><span class="pre">pretrain.py</span></code>. For instance:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">config-path</span></code>: Refers to the configuration folder, e.g., <code class="docutils literal notranslate"><span class="pre">examples/protein/esm2nv/conf</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">config-name</span></code>: Refers to the YAML configuration file, e.g., <code class="docutils literal notranslate"><span class="pre">pretrain_esm2_650M.yaml</span></code>.</p></li>
</ul>
<p>The full path for the configuration file in this example would be:<br />
<code class="docutils literal notranslate"><span class="pre">examples/protein/esm2nv/conf/pretrain_esm2_650M.yaml</span></code>.</p>
</li>
</ul>
<p>Parameters starting with <code class="docutils literal notranslate"><span class="pre">++</span></code> are configurable within the YAML file. Below are some examples of such parameters found in the <code class="docutils literal notranslate"><span class="pre">pretrain_esm2_650M.yaml</span></code> file, which inherits from <code class="docutils literal notranslate"><span class="pre">base_config.yaml</span></code>:</p>
<ul class="simple">
<li><p><strong><code class="docutils literal notranslate"><span class="pre">do_training</span></code></strong>:<br />
Set to <code class="docutils literal notranslate"><span class="pre">False</span></code> if you only want to preprocess the data without initiating training.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">model.data.val_size</span></code></strong> and <strong><code class="docutils literal notranslate"><span class="pre">model.data.test_size</span></code></strong>:<br />
These specify the sizes of the validation and test datasets, respectively.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">model.data.train.uf50_datapath</span></code></strong>:<br />
Specifies the path to the UniRef50 FASTA file.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">model.data.train.uf90_datapath</span></code></strong>:<br />
Specifies the path to the UniRef90 FASTA file.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">model.data.train.cluster_mapping_tsv</span></code></strong>:<br />
Specifies the path to the mapping file that maps UniRef50 clusters to UniRef90 sequences.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">model.data.dataset_path</span></code></strong>:<br />
Specifies the path to the output directory for the preprocessed UniRef50 and UniRef90 data. After processing, the following directories will be created:</p>
<ul>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">uf50</span></code></strong>:<br />
Contains <code class="docutils literal notranslate"><span class="pre">train</span></code>/<code class="docutils literal notranslate"><span class="pre">test</span></code>/<code class="docutils literal notranslate"><span class="pre">val</span></code> splits, each with files like <code class="docutils literal notranslate"><span class="pre">x000.csv</span></code>.</p></li>
<li><p><strong><code class="docutils literal notranslate"><span class="pre">uf90</span></code></strong>:<br />
Contains a folder named <code class="docutils literal notranslate"><span class="pre">uf90_csvs</span></code>, with files like <code class="docutils literal notranslate"><span class="pre">x000.csv</span></code>. Note that there will be no train/test/val splits in this directory, as UniRef90 is only used during training.</p></li>
</ul>
</li>
</ul>
<p>Changes can also be made directly to the YAML file instead of overwriting arguments through the command line.</p>
</div>
</div>
<div class="section" id="pretrain-from-scratch">
<h2>Pretrain from Scratch<a class="headerlink" href="#pretrain-from-scratch" title="Permalink to this headline">#</a></h2>
<p>Now we will perform pretraining of ESM-2 from scratch using our prepared data and the parameters provided in the <code class="docutils literal notranslate"><span class="pre">pretrain_esm2_650M.yaml</span></code> config file located in the <code class="docutils literal notranslate"><span class="pre">${bionemo_home}/examples/protein/esm2nv/conf</span></code> folder.</p>
<p>For the purpose of this demo example, we will shorten the time required for training by setting the following parameters: <code class="docutils literal notranslate"><span class="pre">++trainer.max_steps=1</span></code> and <code class="docutils literal notranslate"><span class="pre">++val_check_interval=1</span></code>. Users can update these parameters by editing the <code class="docutils literal notranslate"><span class="pre">.yaml</span></code> config file or by overriding config arguments at runtime using Hydra, as shown in the example below.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">trainer.devices</span></code>: Specifies the number of GPUs to use.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">trainer.max_steps</span></code>: Sets the maximum number of training steps.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">trainer.val_check_interval</span></code>: Determines how often to run validation.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">trainer.limit_train_batches</span></code> and <code class="docutils literal notranslate"><span class="pre">trainer.limit_val_batches</span></code>: Limit the number of batches for training and validation respectively.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">micro_batch_size</span></code>: Refers to the number of samples processed in a single forward/backward pass before performing a weight update.</p></li>
</ul>
<p>Lastly, you can change the configuration used to <code class="docutils literal notranslate"><span class="pre">pretrain_esm2_8M</span></code> if you have hardware constraints.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
! python examples/protein/esm2nv/pretrain.py \
  --config-path=conf \
  --config-name=pretrain_esm2_650M \
  name={model_name}_pretrain \
  ++do_training=True \
  ++model.data.dataset_path={datapath_dir} \
  ++trainer.devices=1 \
  ++model.micro_batch_size=1 \
  ++trainer.max_steps=1 \
  ++trainer.val_check_interval=1 \
  ++exp_manager.create_wandb_logger=False \
  ++trainer.limit_train_batches=1 \
  ++trainer.limit_val_batches=1
</pre></div>
</div>
</div>
</div>
<p>As the ESM-2nv model training job is launched, BioNeMo will print out details related to <strong>compute resources</strong>, <strong>model training configuration</strong>, and the <strong>dataset</strong> being used for training. As the job progresses, it will also print various details related to the test/train/validation steps and accuracy metrics at set intervals.</p>
<p>Upon completion of the training process, it will print details related to log files, model checkpoints, and so on, which will also be saved in the directory as configured (usually <code class="docutils literal notranslate"><span class="pre">/result</span></code>).</p>
<p>Finally, if Weights and Biases logging was enabled (for example, <code class="docutils literal notranslate"><span class="pre">++exp_manager.create_wandb_logger=True</span></code>), you can visualize the model training progress and resulting metrics. The summary will also be printed in the terminal at the end of the training job.</p>
</div>
<div class="section" id="continue-training-from-an-existing-checkpoint">
<h2>Continue Training from an Existing Checkpoint<a class="headerlink" href="#continue-training-from-an-existing-checkpoint" title="Permalink to this headline">#</a></h2>
<p>To continue the pretraining of the foundation model, use the <code class="docutils literal notranslate"><span class="pre">pretrain.py</span></code> script and set <code class="docutils literal notranslate"><span class="pre">exp_manager.resume_if_exists=True</span></code> to load the model weights and previous run’s metadata. This configuration also picks up the learning rate from the scheduler at the end of the previous run.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span>%%capture --no-display --no-stderr cell_output
! python examples/protein/esm2nv/pretrain.py \
  --config-path=conf \
  --config-name=pretrain_esm2_650M \
  name={model_name}_continued_training \
  ++do_training=True \
  ++model.data.dataset_path={datapath_dir} \
  ++trainer.devices=1 \
  ++model.micro_batch_size=1 \
  ++trainer.max_steps=2 \
  ++trainer.val_check_interval=1 \
  ++exp_manager.create_wandb_logger=False \
  ++exp_manager.resume_if_exists=True \
  ++trainer.limit_train_batches=1 \
  ++trainer.limit_val_batches=1
</pre></div>
</div>
</div>
</div>
<p>If Weights and Biases logging was enabled (for example, <code class="docutils literal notranslate"><span class="pre">++exp_manager.create_wandb_logger=True</span></code>), you can also visualize the model training progress and resulting metrics. The summary will also be printed in the terminal at the end of the training job.</p>
<p>In other notebooks, you can explore how to <a class="reference internal" href="esm2_oas_inferencing.html"><span class="doc std std-doc">perform inference on your own data</span></a>, <span class="xref myst">cluster such embeddings</span>, <span class="xref myst">bring and preprocess your own data for training your own ESM model</span>, and <span class="xref myst">finetune existing ESM models</span>.</p>
</div>
</div>

<div class="section">
   
</div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="model_training_esm1nv.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">ESM1nv Model Training using BioNeMo</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="model_training_mmb.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">MegaMolBART Model Training using BioNeMo</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By NVIDIA<br/>
  
      &copy; Copyright 2023-2024 NVIDIA CORPORATION &amp; AFFILIATES. All rights reserved..<br/>
    Last updated on Oct 17, 2024.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>