
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Generating Embeddings for Protein Clustering &#8212; NVIDIA BioNeMo Framework</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="https://js.hcaptcha.com/1/api.js"></script>
    <script src="//assets.adobedtm.com/5d4962a43b79/c1061d2c5e7b/launch-191c2462b890.min.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script src="../_static/design-tabs.js"></script>
    <link rel="canonical" href="https://docs.nvidia.com/bionemo-framework/0.4.0/notebooks/protein-esm2nv-clustering.html" />
    <link rel="shortcut icon" href="../_static/nvidia-logo-vert-rgb-blk-for-screen.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART" href="retrosynthesis-notebook.html" />
    <link rel="prev" title="Finetuning LLM in BioNeMo for a Downstream Task" href="physchem-notebook-fw.html" /> 
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
      
<style type="text/css">
  ul.ablog-archive {
    list-style: none;
    overflow: auto;
    margin-left: 0px;
  }
  ul.ablog-archive li {
    float: left;
    margin-right: 5px;
    font-size: 80%;
  }
  ul.postlist a {
    font-style: italic;
  }
  ul.postlist-style-disc {
    list-style-type: disc;
  }
  ul.postlist-style-none {
    list-style-type: none;
  }
  ul.postlist-style-circle {
    list-style-type: circle;
  }
</style>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/nvidia-logo-horiz-rgb-blk-for-screen.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">NVIDIA BioNeMo Framework</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GETTING STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../index.html">
   What is BioNeMo?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../pre-reqs.html">
   Hardware and Software Prerequisites
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../access-startup.html">
   Access and Startup
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../initialization-guide.html">
   Initialization Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../tutorial-list.html">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BioNeMo CORE
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../bionemo-fw-for-model-training-fw.html">
   Fundamentals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../data-module-fw.html">
   Data Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../dwnstr-task-validation.html">
   Validation with a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../inference-triton-fw.html">
   Inference with Nvidia Triton Server
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../deep-dive-esm1-pytriton-inference.html">
   Example of PyTriton Inference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BEST PRACTICES
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../hyperparameters-fw.html">
   Hyperparameter Usage and Tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../parallelism-fw.html">
   Training Parallelism
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  MODELS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../models/diffdock.html">
   DiffDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/dnabert.html">
   DNABERT
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/dsmbind.html">
   Model Overview
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/equidock.html">
   EquiDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/esm1-nv.html">
   ESM-1nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/esm2-nv.html">
   ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/geneformer.html">
   Geneformer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/megamolbart.html">
   MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/molmim.html">
   MolMIM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/openfold.html">
   OpenFold
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/prott5nv.html">
   Prott5nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../models/model-benchmarks.html">
   Model Benchmarks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  DATASETS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/uniprot.html">
   UniProt Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/CELLxGENE.html">
   CELLxGENE
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/zinc15.html">
   ZINC-15 Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/pdb.html">
   The Protein Data Bank (PDB)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/GRCh38.p13.html">
   Human Reference Genome Version GRCh38.p13
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/openproteinset.html">
   OpenProteinSet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../datasets/moleculenet-physchem.html">
   MoleculeNet Physical Chemistry Datasets - Lipophilicity, FreeSolv, ESOL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  TUTORIALS
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing-bcp-training-diffdock.html">
   DiffDock: Preparing Workspace and Data for Pre-training
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2nv-mutant-design.html">
   Zero-Shot Protein Design Using ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MMB_GenerativeAI_Inference_with_examples.html">
   BioNeMo - MegaMolBART Inferencing for Generative Chemistry
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="MolMIM_GenerativeAI_local_inference_with_examples.html">
   MolMIM Inferencing for Generative Chemistry and Downstream Prediction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="ZINC15-data-preprocessing.html">
   ZINC Training Dataset Setup for small molecule language models
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bionemo-finetuning-overview.html">
   Finetune pre-trained models in BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cma_es_guided_molecular_optimization_molmim.html">
   MolMIM Property Guided Molecular Optimization Using CMA-ES
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-class-fw.html">
   Adding the OAS Dataset: Modifying the Dataset Class
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-dataloader.html">
   Adding the OAS Dataset: Customizing Dataset Object and Dataloader Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="custom-dataset-preprocessing-fw.html">
   Adding the OAS Dataset: Downloading and Preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="encoder-finetuning-notebook-fw.html">
   Encoder Fine-tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_diffdock.html">
   DiffDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_equidock.html">
   EquiDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_esm1nv.html">
   ESM1nv Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_esm2nv.html">
   ESM-2nv: Data Preprocessing and Model Training Using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_mmb.html">
   MegaMolBART Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="model_training_molmim.html">
   Train MolMIM from scratch on your own data using the BioNeMo Framework
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="physchem-notebook-fw.html">
   Finetuning LLM in BioNeMo for a Downstream Task
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Generating Embeddings for Protein Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="retrosynthesis-notebook.html">
   Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_FLIP_finetuning.html">
   Fine-Tune ESM-2nv on FLIP Data for Sequence-Level Classification, Regression, Token-Level Classification, and with LoRA Adapters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_oas_inferencing.html">
   Performing inference on OAS sequences with ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="esm2_paratope_finetuning.html">
   Pretrain from Scratch, Continue Training from an Existing Checkpoint, and Fine-tune ESM-2nv on Custom Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dnabert_inference.html">
   Generating and visualizing embeddings with DNABert
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dnabert_pretrain_finetune.html">
   Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  APPENDIX
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../faq-fw.html">
   Frequently Asked Questions (FAQ)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../bibliography-fw.html">
   Bibliography
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../releasenotes-fw.html">
   Release Notes
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Generating Embeddings for Protein Clustering
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#setup">
     Setup
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#launch-a-docker-container">
       Launch a Docker container
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#prepare-data">
       Prepare data
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#download-pre-trained-models">
       Download pre-trained Models
      </a>
      <ul class="visible nav section-nav flex-column">
       <li class="toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#configuring-paths-in-your-active-container">
         Configuring paths in your active container
        </a>
       </li>
       <li class="toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#setting-up-the-notebook-in-your-active-container">
         Setting up the notebook in your active container
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#run-inference-on-protein-sequence-data">
     Run inference on protein sequence data
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#protein-sequence-clustering-example">
     Protein sequence clustering example
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#esm-2nv-limits">
       ESM-2nv limits
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#data-cleaning">
       Data Cleaning
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#conclusion">
   Conclusion
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Generating Embeddings for Protein Clustering</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Generating Embeddings for Protein Clustering
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#setup">
     Setup
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#launch-a-docker-container">
       Launch a Docker container
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#prepare-data">
       Prepare data
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#download-pre-trained-models">
       Download pre-trained Models
      </a>
      <ul class="visible nav section-nav flex-column">
       <li class="toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#configuring-paths-in-your-active-container">
         Configuring paths in your active container
        </a>
       </li>
       <li class="toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#setting-up-the-notebook-in-your-active-container">
         Setting up the notebook in your active container
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#run-inference-on-protein-sequence-data">
     Run inference on protein sequence data
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#protein-sequence-clustering-example">
     Protein sequence clustering example
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#esm-2nv-limits">
       ESM-2nv limits
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#data-cleaning">
       Data Cleaning
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#conclusion">
   Conclusion
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                 <div class="tex2jax_ignore mathjax_ignore section" id="generating-embeddings-for-protein-clustering">
<h1>Generating Embeddings for Protein Clustering<a class="headerlink" href="#generating-embeddings-for-protein-clustering" title="Permalink to this headline">#</a></h1>
<p>In this example we will use the BioNeMo Framework to show how to</p>
<ul class="simple">
<li><p>Set up the essential BioNeMo building blocks for ML inference</p>
<ul>
<li><p>Launch a BioNeMo Docker container</p></li>
<li><p>Prepare data</p></li>
<li><p>Download pretrained models</p></li>
<li><p>Launch an inference server for a pretrained model</p></li>
</ul>
</li>
<li><p>Run inference on protein sequence data</p></li>
<li><p>Walk through a protein sequence clustering task with example data</p>
<ul>
<li><p>Extract embeddings</p></li>
<li><p>Build simple visualizations</p></li>
<li><p>Perform a Clustering task with the embeddings</p></li>
</ul>
</li>
</ul>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">#</a></h2>
<p>Before diving in, please ensure that you have completed all steps in the <a class="reference internal" href="../index.html"><span class="doc std std-doc">Getting Started</span></a> section.</p>
<p>Additionally, this tutorial depends on:
- The <span class="xref myst">ESM-2</span> model.
- The local <a class="reference internal" href="../inference-triton-fw.html"><span class="doc std std-doc">inference server</span></a>.</p>
<div class="section" id="launch-a-docker-container">
<h3>Launch a Docker container<a class="headerlink" href="#launch-a-docker-container" title="Permalink to this headline">#</a></h3>
<p>Once you have successfully set up the BioNeMo Framework Docker image, identify the image ID by running ‘docker images’ and finding the correct BioNeMo image that you’ve downloaded.<br />
With the required model name, model path details, you can now run the container locally with <code class="docutils literal notranslate"><span class="pre">docker</span> <span class="pre">run</span> <span class="pre">-it</span> <span class="pre">-p</span> <span class="pre">8888:8888</span> <span class="pre">--gpus</span> <span class="pre">all</span> <span class="pre">-v</span> <span class="pre">&lt;model</span> <span class="pre">directory&gt;:&lt;mount</span> <span class="pre">destination&gt;</span> <span class="pre">&lt;BioNeMo_Image&gt;</span> <span class="pre">&lt;model&gt;</span></code>. This command will set up a ‘Jupyter Lab’ that can be found on <code class="docutils literal notranslate"><span class="pre">localhost:8888</span></code>.</p>
<p><em>This example has been built upon a container running in a local machine with 2 x A6000 RTX GPUs.</em></p>
</div>
<div class="section" id="prepare-data">
<h3>Prepare data<a class="headerlink" href="#prepare-data" title="Permalink to this headline">#</a></h3>
<p>ESM2-nv can be used to run inference on protein sequence data, resulting in embeddings.<br />
In this example, we will work with a sample csv file with sequence data from the UniRef50 dataset. Our sample dataset consists of 5,001 proteins, and is available within the BioNeMo Framework GitHub repository at the following location: <code class="docutils literal notranslate"><span class="pre">./examples/tests/test_data/uniref202104_esm2_qc</span></code></p>
<p>The data is stored in zip file, so run the following command to extract the various pre-training and inference files.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>unzip<span class="w"> </span>-o<span class="w"> </span>/workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc.zip<span class="w"> </span>-d<span class="w"> </span>/workspace/bionemo/examples/tests/test_data/
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Archive:  /workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc.zip
  inflating: /workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc/README.md  
  inflating: /workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc/ur90_ur50_sampler.fasta  
  inflating: /workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc/uniref50_train_filt.fasta  
  inflating: /workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc/mapping.tsv  
  inflating: /workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc/uniref50_inference_example.csv  
</pre></div>
</div>
</div>
</div>
<p>For the purpose of this inference example, we simply use the <code class="docutils literal notranslate"><span class="pre">uniref50_inference_example.csv</span></code> file. This file is a two-column, comma-separated file where the first column is the sequence IDs and the second column is the protein sequences.</p>
</div>
<div class="section" id="download-pre-trained-models">
<h3>Download pre-trained Models<a class="headerlink" href="#download-pre-trained-models" title="Permalink to this headline">#</a></h3>
<p>BioNeMo Framework comes with the basic building blocks for ML inference, namely the NeMo optimized pre-trained ML models and the framework to use these models for pre-training and inference. We also provide a set of scripts to download these models from NVIDIA NGC.<br />
As of the pre-release version (01 NOV 2023), we have the following models available:</p>
<ul class="simple">
<li><p>Molecule/MegaMolBART</p></li>
<li><p>Protein/ESM-1nv</p></li>
<li><p>Protein/ESM-2nv: 650M and 3B variants.</p></li>
<li><p>Protein/ProtT5nv</p></li>
</ul>
<p>In your active Docker container, <strong>download the pre-trained models</strong> with</p>
<p><code class="docutils literal notranslate"><span class="pre">./launch.sh</span> <span class="pre">download</span></code></p>
<p>The command above will invoke a routine that downloads the models from published addresses in <code class="docutils literal notranslate"><span class="pre">download_bionemo_models</span></code>.</p>
<p>Once execution is complete, you will see files with the <code class="docutils literal notranslate"><span class="pre">.nemo</span></code> extension within the folder specified in the <code class="docutils literal notranslate"><span class="pre">DOCKER_MODELS_PATH</span></code> configuration of your <code class="docutils literal notranslate"><span class="pre">.env</span></code> file.</p>
<p>Refer to the <span class="xref myst">Initialization Guide</span> section for configuring your Docker container and for more info about the <code class="docutils literal notranslate"><span class="pre">.env</span></code> file.</p>
<div class="section" id="configuring-paths-in-your-active-container">
<h4>Configuring paths in your active container<a class="headerlink" href="#configuring-paths-in-your-active-container" title="Permalink to this headline">#</a></h4>
<blockquote>
<div><p><strong>.nemo Files</strong><br />
Make sure the models downloaded are within the desired project mount directory such as <code class="docutils literal notranslate"><span class="pre">/models/protein/esm2nv</span></code>; otherwise, you may have to move the <code class="docutils literal notranslate"><span class="pre">.nemo</span></code> files, such as <code class="docutils literal notranslate"><span class="pre">esm2nv_650M_converted.nemo</span></code> from wherever it has been downloaded into.</p>
</div></blockquote>
<blockquote>
<div><p><strong>Jupyter Lab’s Workspace</strong><br />
If you type <code class="docutils literal notranslate"><span class="pre">pwd</span></code> in the Jupyter Lab’s terminal, this may not show BioNeMo’s original directory in your machine but yet another directory such as <code class="docutils literal notranslate"><span class="pre">/workspace</span></code>, according to the project mount directory you’ve set in the <code class="docutils literal notranslate"><span class="pre">.env</span></code> file.</p>
</div></blockquote>
</div>
<div class="section" id="setting-up-the-notebook-in-your-active-container">
<h4>Setting up the notebook in your active container<a class="headerlink" href="#setting-up-the-notebook-in-your-active-container" title="Permalink to this headline">#</a></h4>
<p>In this example, we will be working with the <a class="reference internal" href="../models/esm2-nv.html"><span class="doc std std-doc">ESM2-nv model</span></a> and a subset of the UniRef50 data (5k proteins) available in the BioNeMo framework.</p>
<p>ESM-2nv was developed using the BioNeMo framework. The model uses an architecture called Bidirectional Encoder Representations from Transformers (BERT) and is based on the ESM-2 model</p>
<p>You could run the code as provided in the cells below - into a Jupyter Notebook launched from the JupyterLab interface running inside the container. For that, you should first run the following two steps on your terminal:</p>
<ol class="arabic simple">
<li><p>Copy the notebook to the example directory of ESM2-nv:</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cp<span class="w"> </span>protein-esm2nv-clustering.ipynb<span class="w"> </span>/workspace/bionemo/examples/esm2nv/nbs
</pre></div>
</div>
<ol class="arabic simple" start="2">
<li><p>Launch a gRPC client with the ESM2-nv-650M model:</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python3<span class="w"> </span>-m<span class="w"> </span>bionemo.model.protein.esm1nv.grpc.service<span class="w"> </span>--model<span class="w"> </span>esm2nv_650M
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="run-inference-on-protein-sequence-data">
<h2>Run inference on protein sequence data<a class="headerlink" href="#run-inference-on-protein-sequence-data" title="Permalink to this headline">#</a></h2>
<p>Once the gRPC client is launched with your pre-trained model of choice, the BioNeMo-way is simple to generate embeddings:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">infer</span> <span class="kn">import</span> <span class="n">ESMInferenceWrapper</span>
<span class="c1"># Initalize the Inference Wrapper that will invoke the pre-trained model</span>
<span class="n">esm</span> <span class="o">=</span> <span class="n">ESMInferenceWrapper</span><span class="p">()</span>
<span class="c1"># Work with a set of sequences (these are invalid sequences despite valid amino-acids)</span>
<span class="n">sequence_examples</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;TRANSCENDINGLIMITS&quot;</span><span class="p">,</span> <span class="s2">&quot;MASTERMINDING&quot;</span><span class="p">]</span>
<span class="c1"># Generate Embeddings</span>
<span class="n">embeddings</span> <span class="o">=</span> <span class="n">esm</span><span class="o">.</span><span class="n">seq_to_embedding</span><span class="p">(</span><span class="n">sequence_examples</span><span class="p">)</span>

<span class="c1"># Print the shape of embeddings</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Shape of embeddings (2, 1280): </span><span class="si">{</span><span class="n">embeddings</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Shape of embeddings (2, 1280): torch.Size([2, 1280])
</pre></div>
</div>
</div>
</div>
<p>This requires fewer lines of code than, for example, <a class="reference external" href="https://huggingface.co/Rostlab/prot_t5_xl_uniref50">Rostlab’s Example in Hugging Face</a>. This is because the <code class="docutils literal notranslate"><span class="pre">ESMInferenceWrapper</span></code> BioNeMo class takes care of the tokenization process and handles the configuration of the hidden states.</p>
</div>
<div class="section" id="protein-sequence-clustering-example">
<h2>Protein sequence clustering example<a class="headerlink" href="#protein-sequence-clustering-example" title="Permalink to this headline">#</a></h2>
<p>Since we are now familiar with the basic building blocks of the BioNeMo Framework to generate protein embeddings, let’s tackle and example including data vizualization and clustering with the UniRef50 sample data.</p>
<p>To perform this in an efficient manner, we will utilize open-source CUDA accelerated data science and machiene learning libraries such as cuML, cuDF, cuPY. For more information, check out <a class="reference external" href="https://rapids.ai/">NVIDIA RAPIDS</a>.</p>
<p><a class="reference external" href="https://github.com/rapidsai/cudf">cuDF</a> provides a pandas-like API that will be familiar to data engineers &amp; data scientists, so they can use it to easily accelerate their workflows without going into the details of CUDA programming. Similarly, <a class="reference external" href="https://github.com/rapidsai/cuml">cuML</a> enables data scientists, researchers, and software engineers to run traditional tabular ML tasks on GPUs. In most cases, cuML’s Python API matches the API from scikit-learn.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The following cells containing python code snippets can be directly copied and executed into a Python environment such as a Jupyter notebook running in a BioNemo container.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>which<span class="w"> </span>python
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/bin/python
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="kn">import</span> <span class="nn">cudf</span>                     <span class="c1"># Using cudf from NVIDIA RAPIDS for big dataframes (alternatively, pandas can be used for simple dataframe manipulations)</span>
<span class="kn">import</span> <span class="nn">cupy</span> <span class="k">as</span> <span class="nn">cp</span>               <span class="c1"># To process arrays on the GPU</span>
<span class="kn">import</span> <span class="nn">cuml</span>                     <span class="c1"># We will use DBSCAN for clustering with unkown labels and clusters</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>              <span class="c1"># numpy for simple array manipulations</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span> <span class="c1"># Matplotlib for data vizualization</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>             <span class="c1"># simple dataframe manipulation</span>
<span class="kn">import</span> <span class="nn">torch</span>                    <span class="c1"># To deal with tensors</span>
<span class="kn">from</span> <span class="nn">cuml.manifold</span> <span class="kn">import</span> <span class="n">TSNE</span> <span class="k">as</span> <span class="n">cumlTSNE</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>NOTE! Installing ujson may make loading annotations faster.
CPU times: user 3.76 s, sys: 4.61 s, total: 8.38 s
Wall time: 8.03 s
</pre></div>
</div>
</div>
</div>
<blockquote>
<div><p>For this example, we will use a small dataset of 5k proteins that we share in the BioNeMo FW and that can be found at <code class="docutils literal notranslate"><span class="pre">/examples/tests/test_data/uniref202104_esm2_q</span></code></p>
</div></blockquote>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="n">data_path</span> <span class="o">=</span> <span class="s2">&quot;/workspace/bionemo/examples/tests/test_data/uniref202104_esm2_qc&quot;</span>
<span class="n">sequence_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_path</span><span class="p">,</span> <span class="s2">&quot;uniref50_inference_example.csv&quot;</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 13.8 ms, sys: 0 ns, total: 13.8 ms
Wall time: 12.8 ms
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sequence_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>sequence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>UniRef50_A0A007</td>
      <td>MGYIHTALKSAGFHHVIQVDTPALGLDSEGLRKLLADFEPDLVGVS...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>UniRef50_A0A009DWD5</td>
      <td>MGAKLTNHDPNKMALIESINYFNKYYQQDIKIYEAVLKP</td>
    </tr>
    <tr>
      <th>2</th>
      <td>UniRef50_A0A009DWJ5</td>
      <td>MAMQHVTDRATRMIHHSGRGSQYCSELYQSALRHYGVCPSMTDGGDC</td>
    </tr>
    <tr>
      <th>3</th>
      <td>UniRef50_A0A009DY31</td>
      <td>MEHAREQRVKRTQRDYSFAFKMMVVHEVEKGQITYKQAQAKYGNLSSI</td>
    </tr>
    <tr>
      <th>4</th>
      <td>UniRef50_A0A009DYA3</td>
      <td>MISETCYRYQAKLSDDNTFIAEQLIELTEENPDWSFGLCFSYLQHV...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let’s have a look on the sample dataset…</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="c1"># Define the bin function</span>
<span class="k">def</span> <span class="nf">bin_data</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">step</span><span class="p">):</span>
    <span class="c1"># Create the bins</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">step</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">]</span>

    <span class="c1"># Create the labels</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="n">step</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">step</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;1001+&#39;</span><span class="p">]</span>
    
    <span class="c1"># Cut the Series</span>
    <span class="n">s_cut</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">include_lowest</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">s_cut</span>

<span class="c1"># Let&#39;s create a column to capture the sequence lengths  </span>
<span class="n">sequence_df</span><span class="p">[</span><span class="s1">&#39;sequence_length&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence</span><span class="p">]</span>  

<span class="c1"># we&#39;ll work with the sequences and their lengths</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

<span class="c1"># Bin the data</span>
<span class="n">s_binned</span> <span class="o">=</span> <span class="n">bin_data</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1"># Count the frequency of each bin</span>
<span class="n">s_counts</span> <span class="o">=</span> <span class="n">s_binned</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="c1"># Generate colors from the viridis colormap</span>
<span class="n">colors</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">viridis</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">s_counts</span><span class="p">)))</span>

<span class="c1"># Create a 1x2 grid of subplots</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Create the bar plot on the first subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">s_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">s_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Sequence Lengths - bin size = 100&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Quantity of Residues&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="c1"># Create the pie chart on the second subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">s_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">s_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Residues Lengths Relative to the Dataset&#39;</span><span class="p">)</span>

<span class="c1"># Display the plots</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/fb6c4adf279599edf7a2bc912df60c1b98ad9ab29edfc6de302479b38391f099.png" src="../_images/fb6c4adf279599edf7a2bc912df60c1b98ad9ab29edfc6de302479b38391f099.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 894 ms, sys: 4.04 s, total: 4.93 s
Wall time: 208 ms
</pre></div>
</div>
</div>
</div>
<div class="section" id="esm-2nv-limits">
<h3>ESM-2nv limits<a class="headerlink" href="#esm-2nv-limits" title="Permalink to this headline">#</a></h3>
<p>The current limit for ESM-2nv is 1023 Residues based on the nemo checkpoints, so we will have to decrease the dataset accordingly.</p>
</div>
<div class="section" id="data-cleaning">
<h3>Data Cleaning<a class="headerlink" href="#data-cleaning" title="Permalink to this headline">#</a></h3>
<p>We can also ignore the test and validation sets for this example (since we’re not training a model).</p>
<p>In this example, we will use two groups of proteins:</p>
<ol class="arabic simple">
<li><p>Up to 100 amino acids, which can include functional types such as</p></li>
</ol>
<ul class="simple">
<li><p><strong>Regulatory Peptides</strong>: For example, Insulin, with 51 amino acids.</p></li>
<li><p><strong>Neuropeptides</strong>: Small protein-like molecules used by neurons to communicate with each other. Substance P, a neuropeptide involved in pain perception, contains 11 amino acids.</p></li>
<li><p><strong>Antimicrobial Peptides</strong>: These are an essential part of the innate immune response and have antibiotic properties. An example is defensins, which typically contain 18-45 amino acids.</p></li>
<li><p><strong>Toxins</strong>: Many toxins produced by plants, animals, and bacteria are small peptides. For example, Melittin, the main toxin in bee venom, has 26 amino acids.</p></li>
<li><p><strong>Cytokines</strong>: Some small cytokines, often referred to as chemokines, have less than 100 amino acids. An example is Interleukin-8 (IL-8), a chemokine produced by macrophages, with 99 amino acids.</p></li>
</ul>
<ol class="arabic simple" start="2">
<li><p>Between 400 and 1000 amino acids, with functional types such as</p>
<ul class="simple">
<li><p><strong>G-Protein Coupled Receptors (GPCRs)</strong>: This is a large family of proteins that span the cell membrane and receive signals from outside the cell. GPCRs are involved in a vast number of physiological processes and they typically have around 350-1000 amino acids.</p></li>
<li><p><strong>Ion Channels</strong>: Ion channels are a class of proteins that allow ions to pass through the membrane in response to a signal. The length of these proteins varies. Some are within the 400-1000 amino acid range, while others are up to 4,000 amino acids.</p></li>
<li><p><strong>Cytochrome P450 Enzymes</strong>: These enzymes are involved in the metabolism of various molecules and chemicals within cells. They are typically around 400-500 amino acids long.</p></li>
<li><p><strong>Serine/Threonine Protein Kinases</strong>: This is a large family of enzymes that catalyze the addition of a phosphate group to serine or threonine residues in other proteins. While the length varies significantly, many members of this family are within the 1,000 amino acid range.</p></li>
<li><p><strong>GTPases</strong>: These are a family of hydrolase enzymes that can bind and hydrolyze guanosine triphosphate (GTP). The proteins are critical regulators of a wide variety of cellular processes. Smaller GTPase proteins are 400-500 amino acids in length, with larger GTPases having lengths around 800 amino acids.</p></li>
</ul>
</li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="n">sequence_df</span> <span class="o">=</span> <span class="n">sequence_df</span><span class="p">[(</span><span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&lt;</span> <span class="mi">100</span><span class="p">)</span> <span class="o">|</span> <span class="p">((</span><span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&gt;</span> <span class="mi">400</span><span class="p">)</span> <span class="o">&amp;</span>  <span class="p">(</span><span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&lt;</span> <span class="mi">1000</span><span class="p">))]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Max seq. length (sample dataset)&quot;</span><span class="p">,</span> <span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Length of dataset&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence_df</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Max seq. length (sample dataset) 997
Length of dataset 2323
CPU times: user 702 µs, sys: 1.44 ms, total: 2.14 ms
Wall time: 1.83 ms
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sequence_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>record_name</th>
      <th>sequence</th>
      <th>sequence_length</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>UniRef50_A0A007</td>
      <td>MGYIHTALKSAGFHHVIQVDTPALGLDSEGLRKLLADFEPDLVGVS...</td>
      <td>407</td>
    </tr>
    <tr>
      <th>1</th>
      <td>UniRef50_A0A009DWD5</td>
      <td>MGAKLTNHDPNKMALIESINYFNKYYQQDIKIYEAVLKP</td>
      <td>39</td>
    </tr>
    <tr>
      <th>2</th>
      <td>UniRef50_A0A009DWJ5</td>
      <td>MAMQHVTDRATRMIHHSGRGSQYCSELYQSALRHYGVCPSMTDGGDC</td>
      <td>47</td>
    </tr>
    <tr>
      <th>3</th>
      <td>UniRef50_A0A009DY31</td>
      <td>MEHAREQRVKRTQRDYSFAFKMMVVHEVEKGQITYKQAQAKYGNLSSI</td>
      <td>48</td>
    </tr>
    <tr>
      <th>4</th>
      <td>UniRef50_A0A009DYA3</td>
      <td>MISETCYRYQAKLSDDNTFIAEQLIELTEENPDWSFGLCFSYLQHV...</td>
      <td>50</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="c1"># Define the bin function</span>
<span class="k">def</span> <span class="nf">bin_data</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">step</span><span class="p">):</span>
    <span class="c1"># Create the bins</span>
    <span class="n">bins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">step</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">]</span>

    <span class="c1"># Create the labels</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="n">step</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">step</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;1000+&#39;</span><span class="p">]</span>
    
    <span class="c1"># Cut the Series</span>
    <span class="n">s_cut</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">include_lowest</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">s_cut</span>

<span class="c1"># we&#39;ll work with the sequences and their lengths</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

<span class="c1"># Bin the data</span>
<span class="n">s_binned</span> <span class="o">=</span> <span class="n">bin_data</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">step</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="c1"># Count the frequency of each bin</span>
<span class="n">s_counts</span> <span class="o">=</span> <span class="n">s_binned</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

<span class="c1"># Generate colors from the viridis colormap</span>
<span class="n">colors</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">viridis</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">s_counts</span><span class="p">)))</span>

<span class="c1"># Create a 1x2 grid of subplots</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Create the bar plot on the first subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">s_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">s_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Sequence Lengths - bin size = 100&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Quantity of Residues&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Frequency&#39;</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">45</span><span class="p">)</span>

<span class="c1"># Create the pie chart on the second subplot</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">s_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">s_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.1f%%</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Residues Lengths Relative to the Dataset&#39;</span><span class="p">)</span>

<span class="c1"># Display the plots</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b9f319d787e9917ac257b5e797dd7a488dcebe3a0345c32acedd3d3ee561daae.png" src="../_images/b9f319d787e9917ac257b5e797dd7a488dcebe3a0345c32acedd3d3ee561daae.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 190 ms, sys: 150 ms, total: 340 ms
Wall time: 171 ms
</pre></div>
</div>
</div>
</div>
<p><strong>Using the Inference Wrapper</strong></p>
<p>We will obtain the embeddings for the desired sequences based on the pre-trained models we have previously downloaded.</p>
<p>To avoid memory issues, we’ll submit sequences individually.</p>
<p>In this exampe we are creating a embeddings vectors to be clustered. We know the vector’s dimension by ESM-2nv models specifications.</p>
<p><strong>Defines a Scaling Function</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="c1"># Initalize the Inference Wrapper that will invoke the pre-trained model</span>
<span class="n">esm</span> <span class="o">=</span> <span class="n">ESMInferenceWrapper</span><span class="p">()</span>

<span class="c1"># Defines the number of samples per amino acid group</span>
<span class="n">samples</span> <span class="o">=</span> <span class="mi">500</span>
<span class="c1"># Set the hidden dim for ESM2nv-650M model</span>
<span class="n">hidden_dim</span> <span class="o">=</span> <span class="mi">1280</span>

<span class="c1"># Creates a balanced dataset</span>
<span class="n">indexes_100</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence_df</span><span class="p">[</span><span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&lt;</span> <span class="mi">100</span><span class="p">]),</span><span class="n">samples</span><span class="p">,</span><span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">indexes_400</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence_df</span><span class="p">[</span><span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">&gt;</span> <span class="mi">400</span><span class="p">]),</span><span class="n">samples</span><span class="p">,</span><span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">balanced_samples</span> <span class="o">=</span> <span class="p">[</span><span class="o">*</span><span class="n">indexes_100</span><span class="p">,</span> <span class="o">*</span><span class="n">indexes_400</span><span class="p">]</span>

<span class="c1"># Initializes an embeddings tensor</span>
<span class="n">prot_embeddings</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">balanced_samples</span><span class="p">),</span> <span class="n">hidden_dim</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">torch</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span>

<span class="c1"># Build the embeddings tensor based on mean pooling</span>
<span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">balanced_samples</span><span class="p">):</span>
    <span class="c1"># Obtain embeddings. </span>
    <span class="n">embeddings</span> <span class="o">=</span> <span class="n">esm</span><span class="o">.</span><span class="n">seq_to_embedding</span><span class="p">(</span><span class="n">sequence_df</span><span class="o">.</span><span class="n">sequence</span><span class="o">.</span><span class="n">to_list</span><span class="p">()[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">prot_embeddings</span><span class="p">[</span><span class="n">_</span><span class="p">,:]</span> <span class="o">=</span> <span class="n">embeddings</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 15.4 s, sys: 1.57 s, total: 17 s
Wall time: 1min 15s
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Auxiliary matrix shape (to be used in clustering) must be (2 x samples, 1280)</span>
<span class="n">prot_embeddings</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>torch.Size([1000, 1280])
</pre></div>
</div>
</div>
</div>
<p><strong>DBSCAN</strong></p>
<p>Script extracted from RAPIDS API tutorial.</p>
<p>More information available at: https://docs.rapids.ai/api/cuml/stable/api/#clustering</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="n">ary</span> <span class="o">=</span> <span class="n">cp</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">prot_embeddings</span><span class="p">)</span>
<span class="n">prev_output_type</span> <span class="o">=</span> <span class="n">cuml</span><span class="o">.</span><span class="n">global_settings</span><span class="o">.</span><span class="n">output_type</span>
<span class="n">cuml</span><span class="o">.</span><span class="n">set_global_output_type</span><span class="p">(</span><span class="s1">&#39;cudf&#39;</span><span class="p">)</span>
<span class="n">dbscan_float</span> <span class="o">=</span> <span class="n">cuml</span><span class="o">.</span><span class="n">DBSCAN</span><span class="p">(</span><span class="n">eps</span><span class="o">=</span><span class="mf">.56</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="n">dbscan_float</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">ary</span><span class="p">)</span>
<span class="n">cuml</span><span class="o">.</span><span class="n">set_global_output_type</span><span class="p">(</span><span class="n">prev_output_type</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Num. Of Clusters: &quot;</span><span class="p">,</span> <span class="n">dbscan_float</span><span class="o">.</span><span class="n">labels_</span><span class="o">.</span><span class="n">nunique</span><span class="p">())</span>

<span class="kn">import</span> <span class="nn">matplotlib.ticker</span> <span class="k">as</span> <span class="nn">ticker</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">dbscan_float</span><span class="o">.</span><span class="n">labels_</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>

<span class="c1"># Count the number of samples in each cluster</span>
<span class="n">unique_labels</span><span class="p">,</span> <span class="n">counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Plot the histogram</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">unique_labels</span><span class="p">,</span> <span class="n">counts</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">viridis</span><span class="p">(</span><span class="n">counts</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">counts</span><span class="p">))))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Cluster Labels&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Embedding Samples&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Embeddings per Cluster&#39;</span><span class="p">)</span>

<span class="c1"># Ensure all labels are displayed on x-axis and they are integers</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span>
<span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_major_locator</span><span class="p">(</span><span class="n">ticker</span><span class="o">.</span><span class="n">MaxNLocator</span><span class="p">(</span><span class="n">integer</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">(</span><span class="n">unique_labels</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Num. Of Clusters:  2
</pre></div>
</div>
<img alt="../_images/27e1bb221e55764ddb1884da1659f98491d02d267578c87139532368b079c077.png" src="../_images/27e1bb221e55764ddb1884da1659f98491d02d267578c87139532368b079c077.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 273 ms, sys: 100 ms, total: 373 ms
Wall time: 291 ms
</pre></div>
</div>
</div>
</div>
<p>Using T-SNE to reduce dimensionality so that we can visualize the clusters of data</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>
<span class="c1"># Hyperparameters picked arbitrarily for this example</span>
<span class="n">tsne</span> <span class="o">=</span> <span class="n">cumlTSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">)</span>
<span class="n">tsne_embedding</span> <span class="o">=</span> <span class="n">cudf</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tsne</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">ary</span><span class="p">))</span>

<span class="c1"># Create scatterplot</span>
<span class="n">colors</span> <span class="o">=</span> <span class="n">dbscan_float</span><span class="o">.</span><span class="n">labels_</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">tsne_embedding</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">tsne_embedding</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span> <span class="n">c</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="c1"># Add a colorbar</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s1">&#39;Color scale&#39;</span><span class="p">)</span>
<span class="c1"># Show plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/d8535b575a046595760d7b5e986a83fb7336575059a3d8f3ab05d4732e6da786.png" src="../_images/d8535b575a046595760d7b5e986a83fb7336575059a3d8f3ab05d4732e6da786.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CPU times: user 578 ms, sys: 135 ms, total: 714 ms
Wall time: 562 ms
</pre></div>
</div>
</div>
</div>
<p>The plot above is for the sake of exemplification based on a subset of random proteins and should not be representative of a full application.</p>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="conclusion">
<h1>Conclusion<a class="headerlink" href="#conclusion" title="Permalink to this headline">#</a></h1>
<p>In this example, you saw how to operate inference on ESM2-nv with BioNeMo and integrate it with classifc clustering and dataframe manipulation activities. We also walked through the modular process to set up a BioNeMo Docker container, launch a remote inference client, and generate embeddings.</p>
</div>

<div class="section">
   
</div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="physchem-notebook-fw.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Finetuning LLM in BioNeMo for a Downstream Task</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="retrosynthesis-notebook.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By NVIDIA<br/>
  
      &copy; Copyright 2023-2024 NVIDIA CORPORATION &amp; AFFILIATES. All rights reserved..<br/>
    Last updated on Oct 17, 2024.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>