
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Release Notes &#8212; NVIDIA BioNeMo Framework</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="https://js.hcaptcha.com/1/api.js"></script>
    <script src="//assets.adobedtm.com/5d4962a43b79/c1061d2c5e7b/launch-191c2462b890.min.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script src="_static/design-tabs.js"></script>
    <link rel="canonical" href="https://docs.nvidia.com/bionemo-framework/0.4.0/releasenotes-fw.html" />
    <link rel="shortcut icon" href="_static/nvidia-logo-vert-rgb-blk-for-screen.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Bibliography" href="bibliography-fw.html" /> 
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
      
<style type="text/css">
  ul.ablog-archive {
    list-style: none;
    overflow: auto;
    margin-left: 0px;
  }
  ul.ablog-archive li {
    float: left;
    margin-right: 5px;
    font-size: 80%;
  }
  ul.postlist a {
    font-style: italic;
  }
  ul.postlist-style-disc {
    list-style-type: disc;
  }
  ul.postlist-style-none {
    list-style-type: none;
  }
  ul.postlist-style-circle {
    list-style-type: circle;
  }
</style>

  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/nvidia-logo-horiz-rgb-blk-for-screen.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">NVIDIA BioNeMo Framework</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  GETTING STARTED
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="index.html">
   What is BioNeMo?
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pre-reqs.html">
   Hardware and Software Prerequisites
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="access-startup.html">
   Access and Startup
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="initialization-guide.html">
   Initialization Guide
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="tutorial-list.html">
   BioNeMo Framework Tutorials
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BioNeMo CORE
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="bionemo-fw-for-model-training-fw.html">
   Fundamentals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="data-module-fw.html">
   Data Module
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dwnstr-task-validation.html">
   Validation with a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="inference-triton-fw.html">
   Inference with Nvidia Triton Server
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="deep-dive-esm1-pytriton-inference.html">
   Example of PyTriton Inference
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  BEST PRACTICES
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="hyperparameters-fw.html">
   Hyperparameter Usage and Tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="parallelism-fw.html">
   Training Parallelism
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  MODELS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="models/diffdock.html">
   DiffDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/dnabert.html">
   DNABERT
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/dsmbind.html">
   DSMBind
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/equidock.html">
   EquiDock
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/esm1-nv.html">
   ESM-1nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/esm2-nv.html">
   ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/geneformer.html">
   Geneformer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/megamolbart.html">
   MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/molmim.html">
   MolMIM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/openfold.html">
   OpenFold
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/prott5nv.html">
   Prott5nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="models/model-benchmarks.html">
   Model Benchmarks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  DATASETS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/uniprot.html">
   UniProt Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/CELLxGENE.html">
   CELLxGENE
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/zinc15.html">
   ZINC-15 Dataset
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/pdb.html">
   The Protein Data Bank (PDB)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/GRCh38.p13.html">
   Human Reference Genome Version GRCh38.p13
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/openproteinset.html">
   OpenProteinSet
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="datasets/moleculenet-physchem.html">
   MoleculeNet Physical Chemistry Datasets - Lipophilicity, FreeSolv, ESOL
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  TUTORIALS
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="preprocessing-bcp-training-diffdock.html">
   DiffDock: Preparing Workspace and Data for Pre-training
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2nv-mutant-design.html">
   Zero-Shot Protein Design Using ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/MMB_GenerativeAI_Inference_with_examples.html">
   BioNeMo - MegaMolBART Inferencing for Generative Chemistry
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/MolMIM_GenerativeAI_local_inference_with_examples.html">
   MolMIM Inferencing for Generative Chemistry and Downstream Prediction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/ZINC15-data-preprocessing.html">
   ZINC Training Dataset Setup for small molecule language models
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/bionemo-finetuning-overview.html">
   Finetune pre-trained models in BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/cma_es_guided_molecular_optimization_molmim.html">
   MolMIM Property Guided Molecular Optimization Using CMA-ES
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/custom-dataset-class-fw.html">
   Adding the OAS Dataset: Modifying the Dataset Class
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/custom-dataset-dataloader.html">
   Adding the OAS Dataset: Customizing Dataset Object and Dataloader Functions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/custom-dataset-preprocessing-fw.html">
   Adding the OAS Dataset: Downloading and Preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/encoder-finetuning-notebook-fw.html">
   Encoder Fine-tuning
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_diffdock.html">
   DiffDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_equidock.html">
   EquiDock Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_esm1nv.html">
   ESM1nv Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_esm2nv.html">
   ESM-2nv: Data Preprocessing and Model Training Using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_mmb.html">
   MegaMolBART Model Training using BioNeMo
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/model_training_molmim.html">
   Train MolMIM from scratch on your own data using the BioNeMo Framework
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/physchem-notebook-fw.html">
   Finetuning LLM in BioNeMo for a Downstream Task
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/protein-esm2nv-clustering.html">
   Generating Embeddings for Protein Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/retrosynthesis-notebook.html">
   Training LLM for a Custom Downstram Task: Retrosynthesis Prediction using MegaMolBART
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2_FLIP_finetuning.html">
   Fine-Tune ESM-2nv on FLIP Data for Sequence-Level Classification, Regression, Token-Level Classification, and with LoRA Adapters
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2_oas_inferencing.html">
   Performing inference on OAS sequences with ESM-2nv
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/esm2_paratope_finetuning.html">
   Pretrain from Scratch, Continue Training from an Existing Checkpoint, and Fine-tune ESM-2nv on Custom Data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/dnabert_inference.html">
   Generating and visualizing embeddings with DNABert
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="notebooks/dnabert_pretrain_finetune.html">
   Pretrain, Fine-tune, and Perform Inference with DNABERT for Splice Site Prediction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  APPENDIX
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="faq-fw.html">
   Frequently Asked Questions (FAQ)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bibliography-fw.html">
   Bibliography
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Release Notes
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-9">
   BioNeMo Framework v1.9
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-features">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#bug-fixes-and-improvements">
     Bug fixes and Improvements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-8">
   BioNeMo Framework v1.8
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#known-issues">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-7">
   BioNeMo Framework v1.7
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-models">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id4">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-6">
   BioNeMo Framework v1.6
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id5">
     New Features
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-5">
   BioNeMo Framework v1.5
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id6">
     New Models
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-4">
   BioNeMo Framework v1.4
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id7">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id8">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id9">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id10">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-3">
   BioNeMo Framework v1.3
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id11">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id12">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id13">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id14">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-2">
   BioNeMo Framework v1.2
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id15">
   New Models
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id16">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id17">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id18">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-1">
   BioNeMo Framework v1.1
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id19">
   New Models
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id20">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id21">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#security-notice">
     Security Notice
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-0">
   BioNeMo Framework v1.0
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id22">
   New Models
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id23">
     New Features
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v0-4-0">
   BioNeMo Framework v0.4.0
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id24">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id25">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id26">
     Known Issues
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-apis">
     New APIs
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Release Notes</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-9">
   BioNeMo Framework v1.9
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-features">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#bug-fixes-and-improvements">
     Bug fixes and Improvements
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-8">
   BioNeMo Framework v1.8
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#known-issues">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-7">
   BioNeMo Framework v1.7
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-models">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id3">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id4">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-6">
   BioNeMo Framework v1.6
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id5">
     New Features
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-5">
   BioNeMo Framework v1.5
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id6">
     New Models
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-4">
   BioNeMo Framework v1.4
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id7">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id8">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id9">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id10">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-3">
   BioNeMo Framework v1.3
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id11">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id12">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id13">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id14">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-2">
   BioNeMo Framework v1.2
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id15">
   New Models
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id16">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id17">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id18">
     Known Issues
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-1">
   BioNeMo Framework v1.1
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id19">
   New Models
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id20">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id21">
     Bug fixes and Improvements
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#security-notice">
     Security Notice
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v1-0">
   BioNeMo Framework v1.0
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#id22">
   New Models
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id23">
     New Features
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bionemo-framework-v0-4-0">
   BioNeMo Framework v0.4.0
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id24">
     New Models
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id25">
     New Features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id26">
     Known Issues
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#new-apis">
     New APIs
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                 <div class="tex2jax_ignore mathjax_ignore section" id="release-notes">
<h1>Release Notes<a class="headerlink" href="#release-notes" title="Permalink to this headline">#</a></h1>
<div class="section" id="bionemo-framework-v1-9">
<h2>BioNeMo Framework v1.9<a class="headerlink" href="#bionemo-framework-v1-9" title="Permalink to this headline">#</a></h2>
<div class="section" id="new-features">
<h3>New Features<a class="headerlink" href="#new-features" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>[Documentation] Updated, executable ESM-2nv notebooks demonstrating: Data preprocessing and model training with custom datasets, Fine-tuning on FLIP data, Inference on OAS sequences, Pre-training from scratch and continuing training</p></li>
<li><p>[Documentation] New notebook demonstrating Zero-Shot Protein Design Using ESM-2nv. Thank you to <a class="reference external" href="https://github.com/awlange">&#64;awlange</a> from A-Alpha Bio for contributing the original version of this recipe!</p></li>
</ul>
</div>
<div class="section" id="bug-fixes-and-improvements">
<h3>Bug fixes and Improvements<a class="headerlink" href="#bug-fixes-and-improvements" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>[Geneformer] Fixed bug in preprocessing due to a relocation of dependent artifacts.</p></li>
<li><p>[Geneformer] Fixes bug in finetuning to use the newer preprocessing constructor.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-8">
<h2>BioNeMo Framework v1.8<a class="headerlink" href="#bionemo-framework-v1-8" title="Permalink to this headline">#</a></h2>
<div class="section" id="id1">
<h3>New Features<a class="headerlink" href="#id1" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>[Documentation] Updated, executable MolMIM notebooks demonstrating: Training on custom data, Inference and downstream prediction, ZINC15 dataset preprocesing, and CMA-ES optimization</p></li>
<li><p>[Dependencies] Upgraded the framework to <a class="reference external" href="https://github.com/NVIDIA/NeMo/tree/v1.23.0">NeMo v1.23</a>, which updates PyTorch to version 2.2.0a0+81ea7a4 and CUDA to version 12.3.</p></li>
</ul>
</div>
<div class="section" id="id2">
<h3>Bug fixes and Improvements<a class="headerlink" href="#id2" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>[ESM2] Fixed a bug in gradient accumulation in encoder fine-tuning</p></li>
<li><p>[MegaMolBART] Make MegaMolBART encoder finetuning respect random seed set by user</p></li>
<li><p>[MegaMolBART] Finetuning with val_check_interval=1 bug fix</p></li>
</ul>
</div>
<div class="section" id="known-issues">
<h3>Known Issues<a class="headerlink" href="#known-issues" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Minor training speed regression observed for models DNABERT, Geneformer, MolMIM</p></li>
<li><p>Two known critical CVEs GHSA-cgwc-qvrx-rf7f, GHSA-mr7h-w2qc-ffc2. The vulnerabilities arise within a package that’s installed by lightning by default. We do not use that package in bionemo framework container. we are also unable to remove the package in question as it’s installed as a side-effect of installing lightning.</p></li>
<li><p>Two known High CVEs from pytorch : GHSA-pg7h-5qx3-wjr3, GHSA-5pcm-hx3q-hm94.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-7">
<h2>BioNeMo Framework v1.7<a class="headerlink" href="#bionemo-framework-v1-7" title="Permalink to this headline">#</a></h2>
<div class="section" id="new-models">
<h3>New Models<a class="headerlink" href="#new-models" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/2023.12.10.570461v1">DSMBind</a>, developed under the BioNeMo framework, is a model which can produce comparative values for ranking protein-ligand binding affinities. This release features the capability to perform inference using a newly trained checkpoint.</p></li>
</ul>
</div>
<div class="section" id="id3">
<h3>New Features<a class="headerlink" href="#id3" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>[EquiDock] Remove steric clashes as a post-processing step after equidock inference.</p></li>
<li><p>[Documentation] Updated Getting Started section which sequentially describes prerequisites, BioNeMo Framework access, startup instructions, and next steps.</p></li>
</ul>
</div>
<div class="section" id="id4">
<h3>Known Issues<a class="headerlink" href="#id4" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>There is a known security vulnerability with NLTK that can allow for arbitrary code execution via pickle files that are external assets downloaded via nltk.download() (https://github.com/nltk/nltk/issues/3266). BioNeMo itself does not use this dependency in any way, however parts of NeMo text-to-speech (nemo.collections.tts) does use this vulnerable codepath. Since NeMo is installed in the BioNeMo release containers, users are urged to exercise caution when using  nemo.collections.tts or nltk.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-6">
<h2>BioNeMo Framework v1.6<a class="headerlink" href="#bionemo-framework-v1-6" title="Permalink to this headline">#</a></h2>
<div class="section" id="id5">
<h3>New Features<a class="headerlink" href="#id5" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>[Model Fine-tuning] <code class="docutils literal notranslate"><span class="pre">model.freeze_layers</span></code> fine-tuning config parameter added to freeze a specified number of layers. Thank you to github user <a class="reference external" href="https://github.com/nehap25">&#64;nehap25</a>!</p></li>
<li><p>[ESM2]  Loading pre-trained ESM2 weights and continue pre-training on the MLM objective on a custom FASTA dataset is now supported.</p></li>
<li><p>[OpenFold] MLPerf feature 3.2 bug (mha_fused_gemm) fix has merged.</p></li>
<li><p>[OpenFold] MLPerf feature 3.10 integrated into bionemo framework.</p></li>
<li><p>[DiffDock] Updated data loading module for DiffDock model training, changing from sqlite3 backend to webdataset.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-5">
<h2>BioNeMo Framework v1.5<a class="headerlink" href="#bionemo-framework-v1-5" title="Permalink to this headline">#</a></h2>
<div class="section" id="id6">
<h3>New Models<a class="headerlink" href="#id6" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://www.nature.com/articles/s41586-023-06139-9">Geneformer</a> is out of <strong>Beta</strong> status. This release includes newly trained checkpoints and benchmarks, including a variant based on the publication with 10M parameters, and the largest variant of geneformer publically available to date with 106M parameters.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-4">
<h2>BioNeMo Framework v1.4<a class="headerlink" href="#bionemo-framework-v1-4" title="Permalink to this headline">#</a></h2>
<div class="section" id="id7">
<h3>New Models<a class="headerlink" href="#id7" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><strong>Beta</strong> <a class="reference external" href="https://www.nature.com/articles/s41586-023-06139-9">Geneformer</a> a foundation model for single-cell data that encodes each cell as represented by an ordered list of differentially expressed genes for that cell.</p></li>
</ul>
</div>
<div class="section" id="id8">
<h3>New Features<a class="headerlink" href="#id8" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><strong>Beta</strong> <a class="reference internal" href="notebooks/geneformer_cellxgene_tutorial.html"><span class="doc std std-doc">Geneformer pretraining with custom datasets</span></a></p></li>
<li><p><a class="reference internal" href="lora-finetuning-esm2.html"><span class="doc std std-doc">Low-Rank Adaptation (LoRA) finetuning for ESM2</span></a></p></li>
</ul>
</div>
<div class="section" id="id9">
<h3>Bug fixes and Improvements<a class="headerlink" href="#id9" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="models/openfold.html"><span class="doc std std-doc">OpenFold training improved benchmarks and validation of optimizations</span></a></p></li>
</ul>
</div>
<div class="section" id="id10">
<h3>Known Issues<a class="headerlink" href="#id10" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>BioNeMo Framework v24.04 container is vulnerable to <a class="reference external" href="https://github.com/advisories/GHSA-whh8-fjgc-qp73">GHSA-whh8-fjgc-qp73</a> in onnx 1.14.0. Users are advised not to open untrusted onnx files with this image. Restrict your mount point to minimize directory traversal impact. A fix for this is scheduled in the 24.05 (May) release.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-3">
<h2>BioNeMo Framework v1.3<a class="headerlink" href="#bionemo-framework-v1-3" title="Permalink to this headline">#</a></h2>
<div class="section" id="id11">
<h3>New Models<a class="headerlink" href="#id11" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>MolMIM implementation under BioNeMo framework, <a class="reference external" href="https://arxiv.org/abs/2208.09016">a small molecule model developed at NVIDIA</a> which can be used to produce embeddings and novel molecules.</p></li>
</ul>
</div>
<div class="section" id="id12">
<h3>New Features<a class="headerlink" href="#id12" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://developer.nvidia.com/blog/new-models-molmim-and-diffdock-power-molecule-generation-and-molecular-docking-in-bionemo/">MolMIM</a> re-trained on more data is now available in the framework, and achieves <a class="reference internal" href="models/molmim.html"><span class="doc std std-doc">state of the art performance</span></a>.</p></li>
<li><p><a class="reference internal" href="notebooks/cma_es_guided_molecular_optimization_molmim.html"><span class="doc std std-doc">MolMIM property guided tutorial notebook</span></a> covering property guided optimization using our new framework model.</p></li>
<li><p><a class="reference internal" href="notebooks/model_training_molmim.html"><span class="doc std std-doc">MolMIM training tutorial</span></a> available walking users through either training from scratch or from an existing checkpoint on your own data.</p></li>
<li><p><a class="reference internal" href="notebooks/MolMIM_GenerativeAI_local_inference_with_examples.html"><span class="doc std std-doc">MolMIM tutorial notebook covering molecular sampling and property prediction</span></a> is also now available.</p></li>
<li><p>Numerous optimizations from <a class="reference external" href="https://developer.nvidia.com/blog/optimizing-openfold-training-for-drug-discovery/">NVIDIA’s entry to the MLPerf competition</a> have been added to OpenFold. Documentation and detailed benchmarks are works in progress and will be published in upcoming releases. This release contains the following performance optimizations:</p>
<ul>
<li><p>Fused GEMMs in multi-head attention (MHA)</p></li>
<li><p>Non-blocking data pipeline</p></li>
<li><p>BF16 precision training</p></li>
<li><p>Fused MHA gating</p></li>
<li><p>Inductor Compiled LayerNorm</p></li>
<li><p>OpenAI Triton LayerNorm kernels</p></li>
<li><p>OpenAI Triton MHA</p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="id13">
<h3>Bug fixes and Improvements<a class="headerlink" href="#id13" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>NeMo upgraded to v1.22 (<a class="reference external" href="https://github.com/NVIDIA/NeMo/releases">see NeMo release notes</a>),</p></li>
<li><p>PyTorch Lightning upgraded to 2.0.7</p></li>
<li><p><a class="reference external" href="https://org.ngc.nvidia.com/setup/installers/cli">NGC CLI</a> has been removed from the release container. If users download models from inside the container (via e.g. <code class="docutils literal notranslate"><span class="pre">download_artifacts.py</span></code> or <code class="docutils literal notranslate"><span class="pre">launch.sh</span> <span class="pre">download</span></code>), the NGC CLI will be auto-installed to pull the models from NGC.</p></li>
</ul>
</div>
<div class="section" id="id14">
<h3>Known Issues<a class="headerlink" href="#id14" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>BioNeMo Framework v24.03 container is vulnerable to <a class="reference external" href="https://github.com/advisories/GHSA-whh8-fjgc-qp73">GHSA-whh8-fjgc-qp73</a> in onnx 1.14.0. Users are advised not to open untrusted onnx files with this image. Restrict your mount point to minimize directory traversal impact.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-2">
<h2>BioNeMo Framework v1.2<a class="headerlink" href="#bionemo-framework-v1-2" title="Permalink to this headline">#</a></h2>
</div>
<div class="section" id="id15">
<h2>New Models<a class="headerlink" href="#id15" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>OpenFold implementation under BioNeMo framework, derived from public OpenFold and DeepMind AlphaFold-2.</p></li>
<li><p>DNABERT implementation for computing embeddings for each nucleotide in the input DNA sequence.</p></li>
</ul>
<div class="section" id="id16">
<h3>New Features<a class="headerlink" href="#id16" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Training recipes for DNABERT and OpenFold, including automated data processing and full configuration for training.</p></li>
<li><p>Example tutorials for running inference using OpenFold.</p></li>
<li><p>Splice Prediction downstream task example for DNABERT.</p></li>
<li><p>Wrapper scripts for DNABERT and OpenFold to launch jobs on BCP.</p></li>
</ul>
</div>
<div class="section" id="id17">
<h3>Bug fixes and Improvements<a class="headerlink" href="#id17" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Interface improvements for ESM2 data ingestion and pre-processing. The interface allows for explicit specification of training, validation, and test sets. The user may set <code class="docutils literal notranslate"><span class="pre">config.model.data.default_dataset_path</span></code> to maintain prior behavior, or set <code class="docutils literal notranslate"><span class="pre">config.model.data.train.dataset_path</span></code>, <code class="docutils literal notranslate"><span class="pre">config.model.data.val.dataset_path</span></code>, <code class="docutils literal notranslate"><span class="pre">config.model.data.test.dataset_path</span></code> which may all be unique.</p></li>
</ul>
</div>
<div class="section" id="id18">
<h3>Known Issues<a class="headerlink" href="#id18" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>OpenFold training speed does not yet include <a class="reference external" href="https://blogs.nvidia.com/blog/scaling-ai-training-mlperf/">MLPerf optimizations</a>, and these will be released in the subsequent release.</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v1-1">
<h2>BioNeMo Framework v1.1<a class="headerlink" href="#bionemo-framework-v1-1" title="Permalink to this headline">#</a></h2>
</div>
<div class="section" id="id19">
<h2>New Models<a class="headerlink" href="#id19" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>EquiDock for protein-protein docking pose prediction</p></li>
<li><p>DiffDock for protein-ligand blind docking pose generation</p></li>
</ul>
<div class="section" id="id20">
<h3>New Features<a class="headerlink" href="#id20" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Training recipes for EquiDock and DiffDock, including automated data processing and full configuration for training.</p></li>
<li><p>Accelerated inference and training for DiffDock via fast tensor-product kernels.</p></li>
<li><p>Example tutorials for running inference using EquiDock and DiffDock.</p></li>
<li><p>Recipes for running EquiDock and DiffDock on BCP and Slurm.</p></li>
<li><p>Pipeline parallel supported for ESM-2nv.</p></li>
<li><p>Migration of inference notebooks to using pytriton.</p></li>
</ul>
</div>
<div class="section" id="id21">
<h3>Bug fixes and Improvements<a class="headerlink" href="#id21" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Faster pre-processing of data on BCP.</p></li>
<li><p>Refactor of download_models.sh to download_models.py for easier CLI use.</p></li>
<li><p>Refactor of install structure to move from /opt/nvidia to /workspace/bionemo. The environment variable $BIONEMO_HOME now points to the repo base and is required to be set for tests to pass.</p></li>
</ul>
</div>
<div class="section" id="security-notice">
<h3>Security Notice<a class="headerlink" href="#security-notice" title="Permalink to this headline">#</a></h3>
<p>SchedMD Slurm in the release container is shipped with a security vulnerability, <a class="reference external" href="https://ubuntu.com/security/CVE-2022-29501">CVE-2022-29501</a>, and therefore this version of Slurm should not be used to run a Slurm cluster (specifically, the processes <code class="docutils literal notranslate"><span class="pre">slurmdbd</span></code>, <code class="docutils literal notranslate"><span class="pre">slurmctld</span></code>, and <code class="docutils literal notranslate"><span class="pre">slurmd</span></code>.</p>
<p>In general, the BioNeMo Framework release is designed to ship code and an environment that would be executed on local workstations, or deployed on clusters for large scale training jobs. This container is not designed to run as a service with public facing APIs. A full summary of security vulnerabilities can be found <a class="reference external" href="https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara/containers/bionemo-framework/security">here</a>.</p>
</div>
</div>
<div class="section" id="bionemo-framework-v1-0">
<h2>BioNeMo Framework v1.0<a class="headerlink" href="#bionemo-framework-v1-0" title="Permalink to this headline">#</a></h2>
</div>
<div class="section" id="id22">
<h2>New Models<a class="headerlink" href="#id22" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>ESM-2nv for protein sequence representations, pretrained weights of ESM-2 650M and ESM-2 3B converted from HF checkpoint available.</p></li>
</ul>
<div class="section" id="id23">
<h3>New Features<a class="headerlink" href="#id23" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Pre-training recipes for ESM-2nv, including automated data processing and full configuration for training</p></li>
<li><p>Fine-tuning of ESM-2nv with encoder frozen or trainable</p></li>
<li><p>Downstream task finetuning support for single-value classification (e.g. subcellular localization), single-value regression (e.g. meltome) and per-token classification (e.g. secondary structure)</p></li>
<li><p>Validation in loop to evaluate performance on downstream tasks during training</p></li>
<li><p>Example tutorials for pre-training, fine tuning, and downstream tasks</p></li>
</ul>
</div>
</div>
<div class="section" id="bionemo-framework-v0-4-0">
<h2>BioNeMo Framework v0.4.0<a class="headerlink" href="#bionemo-framework-v0-4-0" title="Permalink to this headline">#</a></h2>
<div class="section" id="id24">
<h3>New Models<a class="headerlink" href="#id24" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>ESM-1nv for protein sequence representations, pretrained weights available</p></li>
<li><p>ProtT5nv for protein sequence representation and sequence-to-sequence tasks, pretrained weights available</p></li>
</ul>
</div>
<div class="section" id="id25">
<h3>New Features<a class="headerlink" href="#id25" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Pre-training for all models, including automated data processing and full configuration for training</p></li>
<li><p>Fine-tuning of MegaMolBART, ESM-1nv, and ProtT5nv with encoder frozen or trainable</p></li>
<li><p>Downstream task example applications – secondary structure prediction for ESM-1nv and ProtT5nv, physchem prediction (lipophilicity, FreeSolv, ESOL) and retrosynthesis prediction for MegaMolBART</p></li>
<li><p>Validation in loop to evaluate performance on downstream tasks during training: physchem prediction (MegaMolBART) and secondary structure prediction (ESM-1nv and ProtT5nv).</p></li>
<li><p>Pipeline parallelism supported as a beta feature. Not fully tested.</p></li>
<li><p>Example notebooks for pre-training, fine tuning, and downstream tasks</p></li>
</ul>
</div>
<div class="section" id="id26">
<h3>Known Issues<a class="headerlink" href="#id26" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>Data preprocessing on DGX Cloud is slow. Faster to do it on a local machine.</p></li>
</ul>
</div>
<div class="section" id="new-apis">
<h3>New APIs<a class="headerlink" href="#new-apis" title="Permalink to this headline">#</a></h3>
<ul class="simple">
<li><p>BioNeMoDataModule - Encapsulates dataset instantiation in bionemo models so that many different datasets can be used with the same model</p></li>
<li><p>EncoderFineTuning - Base class to facilitate implementation of downstream tasks built on embeddings from other models</p></li>
</ul>
</div>
</div>
</div>

<div class="section">
   
</div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="bibliography-fw.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Bibliography</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By NVIDIA<br/>
  
      &copy; Copyright 2023-2024 NVIDIA CORPORATION &amp; AFFILIATES. All rights reserved..<br/>
    Last updated on Nov 20, 2024.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>